--- a/net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java
+++ b/net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java
@@ -41,6 +_,30 @@
                                     && !p_260357_.test(livingentity)) {
                                     return true;
                                 } else {
+                                    // Paper start - better track target change reason
+                                    final org.bukkit.event.entity.EntityTargetEvent.TargetReason reason;
+                                    if (!p_258796_.canAttack(livingentity)) {
+                                        reason = org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_INVALID;
+                                    } else if (p_260319_ && StopAttackingIfTargetInvalid.isTiredOfTryingToReachTarget(livingentity, p_258801_.tryGet(p_258788_))) {
+                                        reason = org.bukkit.event.entity.EntityTargetEvent.TargetReason.FORGOT_TARGET;
+                                    } else if (!livingentity.isAlive()) {
+                                        reason = org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_DIED;
+                                    } else if (livingentity.level() != p_258796_.level()) {
+                                        reason = org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_OTHER_LEVEL;
+                                    } else {
+                                        reason = org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_INVALID;
+                                    }
+                                    // Paper end
+                                    // CraftBukkit start
+                                    org.bukkit.event.entity.EntityTargetEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTargetLivingEvent(livingentity, null, reason); // Paper
+                                    if (event.isCancelled()) {
+                                        return false;
+                                    }
+                                    if (event.getTarget() != null) {
+                                        livingentity.getBrain().setMemory(MemoryModuleType.ATTACK_TARGET, ((org.bukkit.craftbukkit.entity.CraftLivingEntity) event.getTarget()).getHandle());
+                                        return true;
+                                    }
+                                    // CraftBukkit end
                                     p_259568_.accept(p_258796_, livingentity);
                                     p_258787_.erase();
                                     return true;
