--- a/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
+++ b/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
@@ -52,6 +_,16 @@
         this.entityGetter = new LevelEntityGetterAdapter<>(this.visibleEntityStorage, this.sectionStorage);
     }
 
+    // CraftBukkit start - add method to get all entities in chunk
+    public List<Entity> getEntities(ChunkPos chunkCoordIntPair) {
+        return sectionStorage.getExistingSectionsInChunk(chunkCoordIntPair.toLong()).flatMap(EntitySection::getEntities).map(entity -> (Entity) entity).collect(Collectors.toList());
+    }
+
+    public boolean isPending(long pair) {
+        return chunkLoadStatuses.get(pair) == ChunkLoadStatus.PENDING;
+    }
+    // CraftBukkit end
+
     void removeSectionIfEmpty(long p_157510_, EntitySection<T> p_157511_) {
         if (p_157511_.isEmpty()) {
             this.sectionStorage.remove(p_157510_);
@@ -59,6 +_,7 @@
     }
 
     private boolean addEntityUuid(T p_157558_) {
+        org.spigotmc.AsyncCatcher.catchOp("Entity add by UUID"); // Paper
         if (!this.knownUuids.add(p_157558_.getUUID())) {
             LOGGER.warn("UUID of added entity already exists: {}", p_157558_);
             return false;
@@ -71,7 +_,17 @@
         return this.addEntity(p_157534_, false);
     }
 
+    public boolean addNewEntityWithoutEvent(T entity) {
+        return this.addEntityWithoutEvent(entity, false);
+    }
+
     private boolean addEntity(T p_157539_, boolean p_157540_) {
+        org.spigotmc.AsyncCatcher.catchOp("Entity add"); // Paper
+        if (p_157539_ instanceof Entity entity && net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.EntityJoinLevelEvent(entity, entity.level(), p_157540_)).isCanceled()) return false;
+        return addEntityWithoutEvent(p_157539_, p_157540_);
+    }
+
+    private boolean addEntityWithoutEvent(T p_157539_, boolean p_157540_) {
         if (!this.addEntityUuid(p_157539_)) {
             return false;
         } else {
@@ -101,27 +_,37 @@
     }
 
     public void addLegacyChunkEntities(Stream<T> p_157553_) {
-        p_157553_.forEach(p_157607_ -> this.addEntity((T)p_157607_, true));
+        p_157553_.forEach(p_157607_ -> {
+            this.addEntity(p_157607_, true);
+            if (p_157607_ instanceof Entity entity) entity.onAddedToLevel();
+        });
     }
 
     public void addWorldGenChunkEntities(Stream<T> p_157560_) {
-        p_157560_.forEach(p_157605_ -> this.addEntity((T)p_157605_, false));
+        p_157560_.forEach(p_157605_ -> {
+            this.addEntity(p_157605_, false);
+            if (p_157605_ instanceof Entity entity) entity.onAddedToLevel();
+        });
     }
 
     void startTicking(T p_157565_) {
+        org.spigotmc.AsyncCatcher.catchOp("Entity start ticking"); // Paper
         this.callbacks.onTickingStart(p_157565_);
     }
 
     void stopTicking(T p_157571_) {
+        org.spigotmc.AsyncCatcher.catchOp("Entity stop ticking"); // Paper
         this.callbacks.onTickingEnd(p_157571_);
     }
 
     void startTracking(T p_157576_) {
+        org.spigotmc.AsyncCatcher.catchOp("Entity start tracking"); // Paper
         this.visibleEntityStorage.add(p_157576_);
         this.callbacks.onTrackingStart(p_157576_);
     }
 
     void stopTracking(T p_157581_) {
+        org.spigotmc.AsyncCatcher.catchOp("Entity stop tracking"); // Paper
         this.callbacks.onTrackingEnd(p_157581_);
         this.visibleEntityStorage.remove(p_157581_);
     }
@@ -132,6 +_,7 @@
     }
 
     public void updateChunkStatus(ChunkPos p_157528_, Visibility p_157529_) {
+        org.spigotmc.AsyncCatcher.catchOp("Update chunk status"); // Paper
         long i = p_157528_.toLong();
         if (p_157529_ == Visibility.HIDDEN) {
             this.chunkVisibility.remove(i);
@@ -165,6 +_,7 @@
     }
 
     public void ensureChunkQueuedForLoad(long p_157556_) {
+        org.spigotmc.AsyncCatcher.catchOp("Entity chunk save"); // Paper
         PersistentEntitySectionManager.ChunkLoadStatus persistententitysectionmanager$chunkloadstatus = this.chunkLoadStatuses.get(p_157556_);
         if (persistententitysectionmanager$chunkloadstatus == PersistentEntitySectionManager.ChunkLoadStatus.FRESH) {
             this.requestChunkLoad(p_157556_);
@@ -172,7 +_,12 @@
     }
 
     private boolean storeChunkSections(long p_157513_, Consumer<T> p_157514_) {
+        return storeChunkSections(p_157513_, p_157514_, false);
+    }
+
+    private boolean storeChunkSections(long p_157513_, Consumer<T> p_157514_, boolean callEvent) {
         PersistentEntitySectionManager.ChunkLoadStatus persistententitysectionmanager$chunkloadstatus = this.chunkLoadStatuses.get(p_157513_);
+
         if (persistententitysectionmanager$chunkloadstatus == PersistentEntitySectionManager.ChunkLoadStatus.PENDING) {
             return false;
         } else {
@@ -180,8 +_,10 @@
                 .getExistingSectionsInChunk(p_157513_)
                 .flatMap(p_157542_ -> p_157542_.getEntities().filter(EntityAccess::shouldBeSaved))
                 .collect(Collectors.toList());
+
             if (list.isEmpty()) {
                 if (persistententitysectionmanager$chunkloadstatus == PersistentEntitySectionManager.ChunkLoadStatus.LOADED) {
+                    if (callEvent) org.bukkit.craftbukkit.event.CraftEventFactory.callEntitiesUnloadEvent(((net.minecraft.world.level.chunk.storage.EntityStorage) permanentStorage).level, new ChunkPos(p_157513_), ImmutableList.of());
                     this.permanentStorage.storeEntities(new ChunkEntities<>(new ChunkPos(p_157513_), ImmutableList.of()));
                 }
 
@@ -190,6 +_,7 @@
                 this.requestChunkLoad(p_157513_);
                 return false;
             } else {
+                if (callEvent) org.bukkit.craftbukkit.event.CraftEventFactory.callEntitiesUnloadEvent(((net.minecraft.world.level.chunk.storage.EntityStorage) permanentStorage).level, new ChunkPos(p_157513_), list.stream().map(entity -> (Entity) entity).collect(Collectors.toList()));
                 this.permanentStorage.storeEntities(new ChunkEntities<>(new ChunkPos(p_157513_), list));
                 list.forEach(p_157514_);
                 return true;
@@ -198,6 +_,7 @@
     }
 
     private void requestChunkLoad(long p_157563_) {
+        org.spigotmc.AsyncCatcher.catchOp("Entity chunk load request"); // Paper
         this.chunkLoadStatuses.put(p_157563_, PersistentEntitySectionManager.ChunkLoadStatus.PENDING);
         ChunkPos chunkpos = new ChunkPos(p_157563_);
         this.permanentStorage.loadEntities(chunkpos).thenAccept(this.loadingInbox::add).exceptionally(p_157532_ -> {
@@ -207,7 +_,11 @@
     }
 
     private boolean processChunkUnload(long p_157569_) {
-        boolean flag = this.storeChunkSections(p_157569_, p_157595_ -> p_157595_.getPassengersAndSelf().forEach(this::unloadEntity));
+        org.spigotmc.AsyncCatcher.catchOp("Entity chunk unload process"); // Paper
+        boolean flag = this.storeChunkSections(p_157569_, (entityaccess) -> {
+            entityaccess.getPassengersAndSelf().forEach(this::unloadEntity);
+        }, true);
+
         if (!flag) {
             return false;
         } else {
@@ -216,24 +_,34 @@
         }
     }
 
-    private void unloadEntity(EntityAccess p_157586_) {
-        p_157586_.setRemoved(Entity.RemovalReason.UNLOADED_TO_CHUNK);
-        p_157586_.setLevelCallback(EntityInLevelCallback.NULL);
+    private void unloadEntity(EntityAccess entityaccess) {
+        entityaccess.setRemoved(Entity.RemovalReason.UNLOADED_TO_CHUNK, org.bukkit.event.entity.EntityRemoveEvent.Cause.UNLOAD);
+        entityaccess.setLevelCallback(EntityInLevelCallback.NULL);
     }
 
     private void processUnloads() {
-        this.chunksToUnload.removeIf(p_157584_ -> this.chunkVisibility.get(p_157584_) != Visibility.HIDDEN ? true : this.processChunkUnload(p_157584_));
+        org.spigotmc.AsyncCatcher.catchOp("Entity chunk process pending loads"); // Paper
+        this.chunksToUnload.removeIf((java.util.function.LongPredicate) (i) -> {
+            return this.chunkVisibility.get(i) != Visibility.HIDDEN ? true : this.processChunkUnload(i);
+        });
     }
 
     private void processPendingLoads() {
         ChunkEntities<T> chunkentities;
+
         while ((chunkentities = this.loadingInbox.poll()) != null) {
-            chunkentities.getEntities().forEach(p_157593_ -> this.addEntity((T)p_157593_, true));
+            chunkentities.getEntities().forEach((entityaccess) -> {
+                this.addEntity(entityaccess, true);
+                if (entityaccess instanceof Entity entity) entity.onAddedToLevel();
+            });
             this.chunkLoadStatuses.put(chunkentities.getPos().toLong(), PersistentEntitySectionManager.ChunkLoadStatus.LOADED);
+            List<Entity> entities = getEntities(chunkentities.getPos());
+            org.bukkit.craftbukkit.event.CraftEventFactory.callEntitiesLoadEvent(((net.minecraft.world.level.chunk.storage.EntityStorage) permanentStorage).level, chunkentities.getPos(), entities);
         }
     }
 
     public void tick() {
+        org.spigotmc.AsyncCatcher.catchOp("Entity manager tick"); // Paper
         this.processPendingLoads();
         this.processUnloads();
     }
@@ -251,6 +_,7 @@
     }
 
     public void autoSave() {
+        org.spigotmc.AsyncCatcher.catchOp("Entity manager autosave"); // Paper
         this.getAllChunksToSave().forEach(p_157579_ -> {
             boolean flag = this.chunkVisibility.get(p_157579_) == Visibility.HIDDEN;
             if (flag) {
@@ -280,7 +_,15 @@
 
     @Override
     public void close() throws IOException {
-        this.saveAll();
+        // CraftBukkit start - add save boolean
+        close(true);
+    }
+
+    public void close(boolean save) throws IOException {
+        if (save) {
+            this.saveAll();
+        }
+        // CraftBukkit end
         this.permanentStorage.close();
     }
 
@@ -367,11 +_,13 @@
 
     class Callback implements EntityInLevelCallback {
         private final T entity;
+        private final Entity realEntity;
         private long currentSectionKey;
         private EntitySection<T> currentSection;
 
         Callback(T p_157614_, long p_157615_, EntitySection<T> p_157616_) {
             this.entity = p_157614_;
+            this.realEntity = p_157614_ instanceof Entity ? (Entity) p_157614_ : null;
             this.currentSectionKey = p_157615_;
             this.currentSection = p_157616_;
         }
@@ -381,6 +_,7 @@
             BlockPos blockpos = this.entity.blockPosition();
             long i = SectionPos.asLong(blockpos);
             if (i != this.currentSectionKey) {
+                org.spigotmc.AsyncCatcher.catchOp("Entity move"); // Paper
                 Visibility visibility = this.currentSection.getStatus();
                 if (!this.currentSection.remove(this.entity)) {
                     PersistentEntitySectionManager.LOGGER
@@ -390,9 +_,11 @@
                 PersistentEntitySectionManager.this.removeSectionIfEmpty(this.currentSectionKey, this.currentSection);
                 EntitySection<T> entitysection = PersistentEntitySectionManager.this.sectionStorage.getOrCreateSection(i);
                 entitysection.add(this.entity);
+                long oldSectionKey = currentSectionKey;
                 this.currentSection = entitysection;
                 this.currentSectionKey = i;
                 this.updateStatus(visibility, entitysection.getStatus());
+                if (this.realEntity != null) net.neoforged.neoforge.common.CommonHooks.onEntityEnterSection(this.realEntity, oldSectionKey, i);
             }
         }
 
@@ -428,6 +_,7 @@
 
         @Override
         public void onRemove(Entity.RemovalReason p_157619_) {
+            org.spigotmc.AsyncCatcher.catchOp("Entity remove"); // Paper
             if (!this.currentSection.remove(this.entity)) {
                 PersistentEntitySectionManager.LOGGER
                     .warn("Entity {} wasn't found in section {} (destroying due to {})", this.entity, SectionPos.of(this.currentSectionKey), p_157619_);
