--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -76,7 +_,7 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.scores.Scoreboard;
 
-public abstract class Level implements LevelAccessor, AutoCloseable, net.neoforged.neoforge.common.extensions.ILevelExtension {
+public abstract class Level extends net.neoforged.neoforge.attachment.AttachmentHolder implements LevelAccessor, AutoCloseable, net.neoforged.neoforge.common.extensions.ILevelExtension {
     public static final Codec<ResourceKey<Level>> RESOURCE_KEY_CODEC = ResourceKey.codec(Registries.DIMENSION);
     public static final ResourceKey<Level> OVERWORLD = ResourceKey.create(Registries.DIMENSION, ResourceLocation.withDefaultNamespace("overworld"));
     public static final ResourceKey<Level> NETHER = ResourceKey.create(Registries.DIMENSION, ResourceLocation.withDefaultNamespace("the_nether"));
@@ -114,7 +_,68 @@
     private final RegistryAccess registryAccess;
     private final DamageSources damageSources;
     private long subTickCount;
-
+    public boolean restoringBlockSnapshots = false;
+    public boolean captureBlockSnapshots = false;
+    public java.util.ArrayList<net.neoforged.neoforge.common.util.BlockSnapshot> capturedBlockSnapshots = new java.util.ArrayList<>();
+    private final java.util.ArrayList<BlockEntity> freshBlockEntities = new java.util.ArrayList<>();
+    private final java.util.ArrayList<BlockEntity> pendingFreshBlockEntities = new java.util.ArrayList<>();
+
+    // CraftBukkit start Added the following
+    public org.bukkit.craftbukkit.CraftWorld world;
+    public boolean pvpMode;
+    public org.bukkit.generator.ChunkGenerator generator;
+    public boolean preventPoiUpdated = false; // CraftBukkit - SPIGOT-5710
+    public boolean captureBlockStates = false;
+    public boolean captureTreeGeneration = false;
+    public java.util.Map<net.minecraft.core.BlockPos, org.bukkit.craftbukkit.block.CraftBlockState> capturedBlockStates = new java.util.LinkedHashMap<>(); // Paper
+    public java.util.Map<net.minecraft.core.BlockPos, net.minecraft.world.level.block.entity.BlockEntity> capturedTileEntities = new java.util.HashMap<>();
+    public List<net.minecraft.world.entity.item.ItemEntity> captureDrops;
+    public final it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap<org.bukkit.entity.SpawnCategory> ticksPerSpawnCategory = new it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap<>();
+    public boolean populating;
+    public org.spigotmc.SpigotWorldConfig spigotConfig; // Spigot
+    // Paper start - add paper world config
+    public io.papermc.paper.configuration.WorldConfiguration paperConfig;
+    public io.papermc.paper.configuration.WorldConfiguration paperConfig() {
+        return this.paperConfig;
+    }
+    public java.util.function.Function<org.spigotmc.SpigotWorldConfig, io.papermc.paper.configuration.WorldConfiguration> paperConfigFunction;
+    // Paper end - add paper world config
+
+    public co.aikar.timings.WorldTimingsHandler timings; // Paper
+    public static BlockPos lastPhysicsProblem; // Spigot
+    protected org.spigotmc.TickLimiter entityLimiter; // Spigot
+    protected org.spigotmc.TickLimiter tileLimiter; // Spigot
+    private int tileTickPosition; // Spigot
+    public final java.util.Map<net.minecraft.world.level.Explosion.CacheKey, Float> explosionDensityCache = new java.util.HashMap<>(); // Paper - Optimize explosions
+    public java.util.ArrayDeque<net.minecraft.world.level.block.RedstoneTorchBlock.Toggle> redstoneUpdateInfos; // Paper - Faster redstone torch rapid clock removal; Move from Map in BlockRedstoneTorch to here
+
+    public org.bukkit.craftbukkit.CraftWorld getWorld() {
+        return this.world;
+    }
+    public org.bukkit.craftbukkit.CraftServer getCraftServer() {
+        return (org.bukkit.craftbukkit.CraftServer) org.bukkit.Bukkit.getServer();
+    }
+
+    public ResourceKey<net.minecraft.world.level.dimension.LevelStem> getTypeKey() {
+        return Registries.levelToLevelStem(dimension);
+    }
+    // CraftBukkit end
+
+    // Magma start
+    protected static org.bukkit.generator.ChunkGenerator chunkGenerator;
+    protected static org.bukkit.generator.BiomeProvider biomeProvider;
+    protected static org.bukkit.World.Environment environment;
+    public String worldName;
+
+
+    public static void setCBWorldInfo(org.bukkit.generator.ChunkGenerator generator, org.bukkit.generator.BiomeProvider biome, org.bukkit.World.Environment env) {
+       chunkGenerator = generator;
+       biomeProvider = biome;
+       environment = env;
+    }
+    // Magma end
+
+    //Magma start - fix dummy level NPE
     protected Level(
         WritableLevelData p_270739_,
         ResourceKey<Level> p_270683_,
@@ -126,6 +_,14 @@
         long p_270248_,
         int p_270466_
     ) {
+        // CraftBukkit Ticks things
+        for (org.bukkit.entity.SpawnCategory spawnCategory : org.bukkit.entity.SpawnCategory.values()) {
+            if (org.bukkit.craftbukkit.util.CraftSpawnCategory.isValidForLimits(spawnCategory)) {
+                this.ticksPerSpawnCategory.put(spawnCategory, (long) this.getCraftServer().getTicksPerSpawns(spawnCategory));
+            }
+        }
+        // CraftBukkit end
+
         this.profiler = p_270692_;
         this.levelData = p_270739_;
         this.dimensionTypeRegistration = p_270240_;
@@ -136,12 +_,12 @@
             this.worldBorder = new WorldBorder() {
                 @Override
                 public double getCenterX() {
-                    return super.getCenterX() / dimensiontype.coordinateScale();
+                    return super.getCenterX(); // CraftBukkit
                 }
 
                 @Override
                 public double getCenterZ() {
-                    return super.getCenterZ() / dimensiontype.coordinateScale();
+                    return super.getCenterZ(); // CraftBukkit
                 }
             };
         } else {
@@ -154,7 +_,75 @@
         this.neighborUpdater = new CollectingNeighborUpdater(this, p_270466_);
         this.registryAccess = p_270200_;
         this.damageSources = new DamageSources(p_270200_);
-    }
+        if (environment == null) {
+            environment = org.magmafoundation.magma.neoforge.NeoInject.environments.get(getTypeKey());
+        }
+        // CraftBukkit start
+        getWorldBorder().world = (net.minecraft.server.level.ServerLevel) this;
+        // From PlayerList.setPlayerFileData
+        getWorldBorder().addListener(new net.minecraft.world.level.border.BorderChangeListener() {
+            @Override
+            public void onBorderSizeSet(WorldBorder worldborder, double d0) {
+                getCraftServer().getHandle().broadcastAll(new net.minecraft.network.protocol.game.ClientboundSetBorderSizePacket(worldborder), worldborder.world);
+            }
+            @Override
+            public void onBorderSizeLerping(WorldBorder worldborder, double d0, double d1, long i) {
+                getCraftServer().getHandle().broadcastAll(new net.minecraft.network.protocol.game.ClientboundSetBorderLerpSizePacket(worldborder), worldborder.world);
+            }
+            @Override
+            public void onBorderCenterSet(WorldBorder worldborder, double d0, double d1) {
+                getCraftServer().getHandle().broadcastAll(new net.minecraft.network.protocol.game.ClientboundSetBorderCenterPacket(worldborder), worldborder.world);
+            }
+            @Override
+            public void onBorderSetWarningTime(WorldBorder worldborder, int i) {
+                getCraftServer().getHandle().broadcastAll(new net.minecraft.network.protocol.game.ClientboundSetBorderWarningDelayPacket(worldborder), worldborder.world);
+            }
+            @Override
+            public void onBorderSetWarningBlocks(WorldBorder worldborder, int i) {
+                getCraftServer().getHandle().broadcastAll(new net.minecraft.network.protocol.game.ClientboundSetBorderWarningDistancePacket(worldborder), worldborder.world);
+            }
+            @Override
+            public void onBorderSetDamagePerBlock(WorldBorder worldborder, double d0) {}
+            @Override
+            public void onBorderSetDamageSafeZOne(WorldBorder worldborder, double d0) {}
+        });
+        // CraftBukkit end
+
+    }
+
+    // Paper start - Cancel hit for vanished players
+    // ret true if no collision
+    public final boolean checkEntityCollision(BlockState data, Entity source, net.minecraft.world.phys.shapes.CollisionContext voxelshapedcollision,
+            BlockPos position, boolean checkCanSee) {
+        // Copied from IWorldReader#a(IBlockData, BlockPosition, VoxelShapeCollision) & EntityAccess#a(Entity, VoxelShape)
+        net.minecraft.world.phys.shapes.VoxelShape voxelshape = data.getCollisionShape(this, position, voxelshapedcollision);
+        if (voxelshape.isEmpty()) {
+            return true;
+        }
+        voxelshape = voxelshape.move((double) position.getX(), (double) position.getY(), (double) position.getZ());
+        if (voxelshape.isEmpty()) {
+            return true;
+        }
+        List<Entity> entities = this.getEntities(null, voxelshape.bounds());
+        for (int i = 0, len = entities.size(); i < len; ++i) {
+            Entity entity = entities.get(i);
+            if (checkCanSee && source instanceof net.minecraft.server.level.ServerPlayer && entity instanceof net.minecraft.server.level.ServerPlayer
+                    && !((net.minecraft.server.level.ServerPlayer) source).getBukkitEntity().canSee(((net.minecraft.server.level.ServerPlayer) entity).getBukkitEntity())) {
+                continue;
+            }
+            // !entity1.dead && entity1.i && (entity == null || !entity1.x(entity));
+            // elide the last check since vanilla calls with entity = null
+            // only we care about the source for the canSee check
+            if (entity.isRemoved() || !entity.blocksBuilding) {
+                continue;
+            }
+            if (net.minecraft.world.phys.shapes.Shapes.joinIsNotEmpty(voxelshape, net.minecraft.world.phys.shapes.Shapes.create(entity.getBoundingBox()), net.minecraft.world.phys.shapes.BooleanOp.AND)) {
+                return false;
+            }
+        }
+        return true;
+    }
+    // Paper end - Cancel hit for vanished players
 
     @Override
     public boolean isClientSide() {
@@ -167,6 +_,13 @@
         return null;
     }
 
+    // Paper start
+    public net.minecraft.world.phys.BlockHitResult.Type clipDirect(Vec3 start, Vec3 end, net.minecraft.world.phys.shapes.CollisionContext context) {
+        // To be patched over
+        return this.clip(new ClipContext(start, end, ClipContext.Block.COLLIDER, ClipContext.Fluid.NONE, context)).getType();
+    }
+    // Paper end
+
     public boolean isInWorldBounds(BlockPos p_46740_) {
         return !this.isOutsideBuildHeight(p_46740_) && isInWorldBoundsHorizontal(p_46740_);
     }
@@ -187,10 +_,65 @@
         return this.getChunk(SectionPos.blockToSectionCoord(p_46746_.getX()), SectionPos.blockToSectionCoord(p_46746_.getZ()));
     }
 
+    @Override
+    public final boolean hasChunkAt(BlockPos pos) {
+        return getChunkIfLoaded(pos.getX() >> 4, pos.getZ() >> 4) != null; // Paper - Perf: Optimize Level.hasChunkAt(BlockPosition)Z
+    }
+
+    public final boolean isLoadedAndInBounds(BlockPos blockposition) { // Paper - final for inline
+        return getWorldBorder().isWithinBounds(blockposition) && getChunkIfLoadedImmediately(blockposition.getX() >> 4, blockposition.getZ() >> 4) != null;
+    }
+
+    public @Nullable LevelChunk getChunkIfLoaded(int x, int z) { // Overridden in WorldServer for ABI compat which has final
+        return ((net.minecraft.server.level.ServerLevel) this).getChunkSource().getChunkAtIfLoadedImmediately(x, z);
+    }
+
+    public final @Nullable LevelChunk getChunkIfLoaded(BlockPos blockposition) {
+        return ((net.minecraft.server.level.ServerLevel) this).getChunkSource().getChunkAtIfLoadedImmediately(blockposition.getX() >> 4, blockposition.getZ() >> 4);
+    }
+
+    //  reduces need to do isLoaded before getType
+    public final @Nullable BlockState getBlockStateIfLoadedAndInBounds(BlockPos blockposition) {
+        return getWorldBorder().isWithinBounds(blockposition) ? getBlockStateIfLoaded(blockposition) : null;
+    }
+
     public LevelChunk getChunk(int p_46727_, int p_46728_) {
         return (LevelChunk)this.getChunk(p_46727_, p_46728_, ChunkStatus.FULL);
     }
 
+    // Paper start - if loaded
+    @Nullable
+    @Override
+    public final ChunkAccess getChunkIfLoadedImmediately(int x, int z) {
+        return ((net.minecraft.server.level.ServerLevel)this).chunkSource.getChunkAtIfLoadedImmediately(x, z);
+    }
+
+    @Override
+    @Nullable
+    public final BlockState getBlockStateIfLoaded(BlockPos pos) {
+        // CraftBukkit start - tree generation
+        if (this.captureTreeGeneration) {
+            org.bukkit.craftbukkit.block.CraftBlockState previous = this.capturedBlockStates.get(pos);
+            if (previous != null) {
+                return previous.getHandle();
+            }
+        }
+        // CraftBukkit end
+        if (this.isOutsideBuildHeight(pos)) {
+            return Blocks.VOID_AIR.defaultBlockState();
+        } else {
+            ChunkAccess chunk = this.getChunkIfLoadedImmediately(pos.getX() >> 4, pos.getZ() >> 4);
+            return chunk == null ? null : chunk.getBlockState(pos);
+        }
+    }
+    // Paper end
+
+    @Override
+    public final FluidState getFluidIfLoaded(BlockPos blockposition) {
+        ChunkAccess chunk = this.getChunkIfLoadedImmediately(blockposition.getX() >> 4, blockposition.getZ() >> 4);
+        return chunk == null ? null : chunk.getFluidState(blockposition);
+    }
+
     @Nullable
     @Override
     public ChunkAccess getChunk(int p_46502_, int p_46503_, ChunkStatus p_331611_, boolean p_46505_) {
@@ -202,13 +_,35 @@
         }
     }
 
+    // Magma start - will allow to call and save for CraftEvent to call the real setblock. this is for mixins
+    public boolean fake$setBlock = false;
+    public BlockState fake$setBlockBlockState = null;
     @Override
     public boolean setBlock(BlockPos p_46601_, BlockState p_46602_, int p_46603_) {
+        if (fake$setBlock) {
+            fake$setBlock = false;
+            fake$setBlockBlockState = p_46602_;
+            return false;
+        }
+        // Magma end
         return this.setBlock(p_46601_, p_46602_, p_46603_, 512);
     }
 
     @Override
     public boolean setBlock(BlockPos p_46605_, BlockState p_46606_, int p_46607_, int p_46608_) {
+        // CraftBukkit start - tree generation
+        if (this.captureTreeGeneration) {
+            org.bukkit.craftbukkit.block.CapturedBlockState blockstate = capturedBlockStates.get(p_46605_);
+            if (blockstate == null) {
+                blockstate = org.bukkit.craftbukkit.block.CapturedBlockState.getTreeBlockState(this, p_46605_, p_46607_);
+                this.capturedBlockStates.put(p_46605_.immutable(), blockstate);
+            }
+            blockstate.setData(p_46606_);
+            blockstate.setFlag(p_46607_);
+            return true;
+        }
+        // CraftBukkit end
+
         if (this.isOutsideBuildHeight(p_46605_)) {
             return false;
         } else if (!this.isClientSide && this.isDebug()) {
@@ -216,11 +_,62 @@
         } else {
             LevelChunk levelchunk = this.getChunkAt(p_46605_);
             Block block = p_46606_.getBlock();
+
+            p_46605_ = p_46605_.immutable(); // Forge - prevent mutable BlockPos leaks
+            net.neoforged.neoforge.common.util.BlockSnapshot blockSnapshot = null;
+            if (this.captureBlockSnapshots && !this.isClientSide) {
+                 blockSnapshot = net.neoforged.neoforge.common.util.BlockSnapshot.create(this.dimension, this, p_46605_, p_46607_);
+                 this.capturedBlockSnapshots.add(blockSnapshot);
+            }
+
+            // CraftBukkit start - capture blockstates
+            boolean captured = false;
+            if (this.captureBlockStates && !this.capturedBlockStates.containsKey(p_46605_)) {
+                org.bukkit.craftbukkit.block.CraftBlockState blockstate = (org.bukkit.craftbukkit.block.CraftBlockState) world.getBlockAt(p_46605_.getX(), p_46605_.getY(), p_46605_.getZ()).getState(); // Paper - use CB getState to get a suitable snapshot
+                blockstate.setFlag(p_46607_); // Paper - set flag
+                this.capturedBlockStates.put(p_46605_.immutable(), blockstate);
+                captured = true;
+            }
+            // CraftBukkit end
+
+            BlockState old = getBlockState(p_46605_);
+            int oldLight = old.getLightEmission(this, p_46605_);
+            int oldOpacity = old.getLightBlock(this, p_46605_);
+
+            levelchunk.setMagma$doPlace((p_46607_ & 1024) == 0);
             BlockState blockstate = levelchunk.setBlockState(p_46605_, p_46606_, (p_46607_ & 64) != 0);
             if (blockstate == null) {
+                if (blockSnapshot != null) this.capturedBlockSnapshots.remove(blockSnapshot);
+                // CraftBukkit start - remove blockstate if failed (or the same)
+                if (this.captureBlockStates && captured) {
+                    this.capturedBlockStates.remove(p_46605_);
+                }
+                // CraftBukkit end
                 return false;
             } else {
                 BlockState blockstate1 = this.getBlockState(p_46605_);
+
+                if (blockSnapshot == null && !this.captureBlockStates) { // Don't notify clients or update physics while capturing blockstates
+                    // Spigot start
+                    try {
+                        this.markAndNotifyBlock(p_46605_, levelchunk, blockstate, p_46606_, p_46607_, p_46608_);
+                    } catch (StackOverflowError ex) {
+                        lastPhysicsProblem = new net.minecraft.core.BlockPos(p_46605_);
+                    }
+                    // Spigot end
+                }
+
+                return true;
+            }
+        }
+    }
+
+    // Split off from original setBlockState(BlockPos, BlockState, int, int) method in order to directly send client and physic updates
+    public void markAndNotifyBlock(BlockPos p_46605_, @Nullable LevelChunk levelchunk, BlockState blockstate, BlockState p_46606_, int p_46607_, int p_46608_) {
+        Block block = p_46606_.getBlock();
+        BlockState blockstate1 = getBlockState(p_46605_);
+        {
+            {
                 if (blockstate1 == p_46606_) {
                     if (blockstate != blockstate1) {
                         this.setBlocksDirty(p_46605_, blockstate, blockstate1);
@@ -247,12 +_,60 @@
                     }
 
                     this.onBlockStateChange(p_46605_, blockstate, blockstate1);
-                }
-
-                return true;
-            }
-        }
-    }
+                    p_46606_.onBlockStateChange(this, p_46605_, blockstate);
+                }
+            }
+        }
+    }
+
+    // CraftBukkit start - Split off from above in order to directly send client and physic updates
+    public void notifyAndUpdatePhysics(BlockPos blockposition, LevelChunk chunk, BlockState oldBlock, BlockState newBlock, BlockState actualBlock, int i, int j) {
+        BlockState iblockdata = newBlock;
+        BlockState iblockdata1 = oldBlock;
+        BlockState iblockdata2 = actualBlock;
+        if (iblockdata2 == iblockdata) {
+            if (iblockdata1 != iblockdata2) {
+                this.setBlocksDirty(blockposition, iblockdata1, iblockdata2);
+            }
+
+            if ((i & 2) != 0 && (!this.isClientSide || (i & 4) == 0) && (this.isClientSide || chunk == null || (chunk.getFullStatus() != null && chunk.getFullStatus().isOrAfter(FullChunkStatus.BLOCK_TICKING)))) { // allow chunk to be null here as chunk.isReady() is false when we send our notification during block placement
+                this.sendBlockUpdated(blockposition, iblockdata1, iblockdata, i);
+            }
+
+            if ((i & 1) != 0) {
+                this.blockUpdated(blockposition, iblockdata1.getBlock());
+                if (!this.isClientSide && iblockdata.hasAnalogOutputSignal()) {
+                    this.updateNeighbourForOutputSignal(blockposition, newBlock.getBlock());
+                }
+            }
+
+            if ((i & 16) == 0 && j > 0) {
+                int k = i & -34;
+
+                // CraftBukkit start
+                iblockdata1.updateIndirectNeighbourShapes(this, blockposition, k, j - 1); // Don't call an event for the old block to limit event spam
+                org.bukkit.craftbukkit.CraftWorld world = ((net.minecraft.server.level.ServerLevel) this).getWorld();
+                if (world != null && ((net.minecraft.server.level.ServerLevel)this).hasPhysicsEvent) { // Paper - BlockPhysicsEvent
+                    org.bukkit.event.block.BlockPhysicsEvent event = new org.bukkit.event.block.BlockPhysicsEvent(world.getBlockAt(blockposition.getX(), blockposition.getY(), blockposition.getZ()), org.bukkit.craftbukkit.block.data.CraftBlockData.fromData(iblockdata));
+                    this.getCraftServer().getPluginManager().callEvent(event);
+
+                    if (event.isCancelled()) {
+                        return;
+                    }
+                }
+                // CraftBukkit end
+                iblockdata.updateNeighbourShapes(this, blockposition, k, j - 1);
+                iblockdata.updateIndirectNeighbourShapes(this, blockposition, k, j - 1);
+            }
+
+            // CraftBukkit start - SPIGOT-5710
+            if (!preventPoiUpdated) {
+                this.onBlockStateChange(blockposition, iblockdata1, iblockdata2);
+            }
+            // CraftBukkit end
+        }
+    }
+    // CraftBukkit end
 
     public void onBlockStateChange(BlockPos p_46609_, BlockState p_46610_, BlockState p_46611_) {
     }
@@ -301,6 +_,7 @@
     }
 
     public void updateNeighborsAt(BlockPos p_46673_, Block p_46674_) {
+        net.neoforged.neoforge.event.EventHooks.onNeighborNotify(this, p_46673_, this.getBlockState(p_46673_), java.util.EnumSet.allOf(Direction.class), false).isCanceled();
     }
 
     public void updateNeighborsAtExceptFromFacing(BlockPos p_46591_, Block p_46592_, Direction p_46593_) {
@@ -342,6 +_,14 @@
 
     @Override
     public BlockState getBlockState(BlockPos p_46732_) {
+        // CraftBukkit start - tree generation
+        if (captureTreeGeneration) {
+            org.bukkit.craftbukkit.block.CraftBlockState previous = this.capturedBlockStates.get(p_46732_); // Paper
+            if (previous != null) {
+                return previous.getHandle();
+            }
+        }
+        // CraftBukkit end
         if (this.isOutsideBuildHeight(p_46732_)) {
             return Blocks.VOID_AIR.defaultBlockState();
         } else {
@@ -489,41 +_,92 @@
         (this.tickingBlockEntities ? this.pendingBlockEntityTickers : this.blockEntityTickers).add(p_151526_);
     }
 
+    public void addFreshBlockEntities(java.util.Collection<BlockEntity> beList) {
+        if (this.tickingBlockEntities) {
+            this.pendingFreshBlockEntities.addAll(beList);
+        } else {
+            this.freshBlockEntities.addAll(beList);
+        }
+    }
+
     protected void tickBlockEntities() {
         ProfilerFiller profilerfiller = this.getProfiler();
         profilerfiller.push("blockEntities");
+        timings.tileEntityPending.startTiming(); // Spigot
+        if (!this.pendingFreshBlockEntities.isEmpty()) {
+            this.freshBlockEntities.addAll(this.pendingFreshBlockEntities);
+            this.pendingFreshBlockEntities.clear();
+        }
         this.tickingBlockEntities = true;
+        if (!this.freshBlockEntities.isEmpty()) {
+            this.freshBlockEntities.forEach(blockEntity -> {
+                // Only call onLoad() on BEs which have been fully added to the level, prevents crashes with BEs that
+                // were discarded due to incompatibility with the BlockState at their position
+                if (!blockEntity.isRemoved() && blockEntity.hasLevel()) {
+                    blockEntity.onLoad();
+                }
+            });
+            this.freshBlockEntities.clear();
+        }
         if (!this.pendingBlockEntityTickers.isEmpty()) {
             this.blockEntityTickers.addAll(this.pendingBlockEntityTickers);
             this.pendingBlockEntityTickers.clear();
         }
+        timings.tileEntityPending.stopTiming(); // Spigot
 
-        Iterator<TickingBlockEntity> iterator = this.blockEntityTickers.iterator();
+        timings.tileEntityTick.startTiming(); // Spigot
+        // Spigot start
+        //Iterator<TickingBlockEntity> iterator = this.blockEntityTickers.iterator();
         boolean flag = this.tickRateManager().runsNormally();
 
-        while (iterator.hasNext()) {
-            TickingBlockEntity tickingblockentity = iterator.next();
+        int tilesThisCycle = 0;
+        var toRemove = new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<TickingBlockEntity>(); // Paper - Fix MC-117075; use removeAll
+        toRemove.add(null); // Paper - Fix MC-117075
+        for (tileTickPosition = 0; tileTickPosition < this.blockEntityTickers.size(); tileTickPosition++) { // Paper - Disable tick limiters
+            tileTickPosition = (tileTickPosition < blockEntityTickers.size()) ? tileTickPosition : 0;
+            TickingBlockEntity tickingblockentity = (TickingBlockEntity) this.blockEntityTickers.get(tileTickPosition);
+            // Spigot end
             if (tickingblockentity.isRemoved()) {
-                iterator.remove();
+                // Spigot start
+                tilesThisCycle--;
+                toRemove.add(tickingblockentity); // Paper - Fix MC-117075; use removeAll
+                // Spigot end
             } else if (flag && this.shouldTickBlocksAt(tickingblockentity.getPos())) {
                 tickingblockentity.tick();
             }
         }
+        this.blockEntityTickers.removeAll(toRemove); // Paper - Fix MC-117075
 
+        timings.tileEntityTick.stopTiming(); // Spigot
         this.tickingBlockEntities = false;
+        co.aikar.timings.TimingHistory.tileEntityTicks += this.blockEntityTickers.size(); // Paper
         profilerfiller.pop();
+        spigotConfig.currentPrimedTnt = 0; // Spigot
     }
 
     public <T extends Entity> void guardEntityTick(Consumer<T> p_46654_, T p_46655_) {
         try {
+            net.neoforged.neoforge.server.timings.TimeTracker.ENTITY_UPDATE.trackStart(p_46655_);
             p_46654_.accept(p_46655_);
         } catch (Throwable throwable) {
-            CrashReport crashreport = CrashReport.forThrowable(throwable, "Ticking entity");
-            CrashReportCategory crashreportcategory = crashreport.addCategory("Entity being ticked");
-            p_46655_.fillCrashReportCategory(crashreportcategory);
-            throw new ReportedException(crashreport);
+            // Paper start - Prevent block entity and entity crashes
+            final String msg = String.format("Entity threw exception at %s:%s,%s,%s", p_46655_.level().getWorld().getName(), p_46655_.getX(), p_46655_.getY(), p_46655_.getZ());
+            MinecraftServer.LOGGER.error(msg, throwable);
+            p_46655_.setRemovedReason(org.bukkit.event.entity.EntityRemoveEvent.Cause.DISCARD); // Magma
+            p_46655_.discard();
+            // Paper end - Prevent block entity and entity crashes
+        } finally {
+            net.neoforged.neoforge.server.timings.TimeTracker.ENTITY_UPDATE.trackEnd(p_46655_);
         }
     }
+
+    // Paper start - Option to prevent armor stands from doing entity lookups
+    @Override
+    public boolean noCollision(@Nullable Entity entity, AABB box) {
+        if (entity instanceof net.minecraft.world.entity.decoration.ArmorStand && !entity.level().paperConfig().entities.armorStands.doCollisionEntityLookups) return false;
+        return LevelAccessor.super.noCollision(entity, box);
+    }
+    // Paper end - Option to prevent armor stands from doing entity lookups
 
     public boolean shouldTickDeath(Entity p_186458_) {
         return true;
@@ -565,20 +_,12 @@
         boolean p_256634_,
         Level.ExplosionInteraction p_256111_
     ) {
-        return this.explode(
-            p_255682_,
-            Explosion.getDefaultDamageSource(this, p_255682_),
-            null,
-            p_255803_,
-            p_256403_,
-            p_256538_,
-            p_255674_,
-            p_256634_,
-            p_256111_,
-            ParticleTypes.EXPLOSION,
-            ParticleTypes.EXPLOSION_EMITTER,
-            SoundEvents.GENERIC_EXPLODE
-        );
+        // Paper start - Allow explosions to damage source
+        return this.explode(p_255682_, p_255803_, p_256403_, p_256538_, p_255674_, p_256634_, p_256111_, null);
+    }
+    public Explosion explode(@Nullable Entity entity, double x, double y, double z, float power, boolean createFire, Level.ExplosionInteraction explosionSourceType, @Nullable Consumer<Explosion> configurator) {
+        return this.explode(entity, Explosion.getDefaultDamageSource(this, entity), (ExplosionDamageCalculator) null, x, y, z, power, createFire, explosionSourceType, ParticleTypes.EXPLOSION, ParticleTypes.EXPLOSION_EMITTER, SoundEvents.GENERIC_EXPLODE, configurator);
+        // Paper end - Allow explosions to damage source
     }
 
     public Explosion explode(
@@ -647,9 +_,13 @@
         ParticleOptions p_311887_,
         Holder<SoundEvent> p_320084_
     ) {
-        return this.explode(
-            p_256233_, p_255861_, p_255867_, p_256447_, p_255732_, p_255717_, p_256013_, p_256228_, p_255784_, true, p_311886_, p_311887_, p_320084_
-        );
+        // Paper start - Allow explosions to damage source
+        return this.explode(p_256233_, p_255861_, p_255867_, p_256447_, p_255732_, p_255717_, p_256013_, p_256228_, p_255784_, p_311886_, p_311887_, p_320084_, null);
+    }
+
+    public Explosion explode(@Nullable Entity entity, @Nullable DamageSource damageSource, @Nullable ExplosionDamageCalculator behavior, double x, double y, double z, float power, boolean createFire, Level.ExplosionInteraction explosionSourceType, ParticleOptions particle, ParticleOptions emitterParticle, Holder<SoundEvent> soundEvent, @Nullable Consumer<Explosion> configurator) {
+        return this.explode(entity, damageSource, behavior, x, y, z, power, createFire, explosionSourceType, true, particle, emitterParticle, soundEvent, configurator);
+        // Paper end - Allow explosions to damage source
     }
 
     public Explosion explode(
@@ -667,14 +_,36 @@
         ParticleOptions p_312060_,
         Holder<SoundEvent> p_320283_
     ) {
+        // Paper start - Allow explosions to damage source
+        return this.explode(p_311934_, p_312790_, p_311975_, p_312493_, p_312456_, p_312719_, p_312292_, p_312144_, p_312265_, p_312145_, p_312842_, p_312060_, null);
+    }
+
+    public Explosion explode(
+            @Nullable Entity p_311934_,
+            @Nullable DamageSource p_312790_,
+            @Nullable ExplosionDamageCalculator p_311975_,
+            double p_312493_,
+            double p_312456_,
+            double p_312719_,
+            float p_312292_,
+            boolean p_312144_,
+            Level.ExplosionInteraction p_312265_,
+            boolean p_312145_,
+            ParticleOptions p_312842_,
+            ParticleOptions p_312060_,
+            Holder<SoundEvent> p_320283_,
+            @Nullable Consumer<Explosion> configurator
+    ) {
+        // Paper end - Allow explosions to damage source
         Explosion.BlockInteraction explosion$blockinteraction = switch (p_312265_) {
             case NONE -> Explosion.BlockInteraction.KEEP;
             case BLOCK -> this.getDestroyType(GameRules.RULE_BLOCK_EXPLOSION_DROP_DECAY);
-            case MOB -> this.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)
+            case MOB -> net.neoforged.neoforge.event.EventHooks.canEntityGrief(this, p_311934_)
             ? this.getDestroyType(GameRules.RULE_MOB_EXPLOSION_DROP_DECAY)
             : Explosion.BlockInteraction.KEEP;
             case TNT -> this.getDestroyType(GameRules.RULE_TNT_EXPLOSION_DROP_DECAY);
             case TRIGGER -> Explosion.BlockInteraction.TRIGGER_BLOCK;
+            case STANDARD -> net.minecraft.world.level.Explosion.BlockInteraction.DESTROY; // CraftBukkit
         };
         Explosion explosion = new Explosion(
             this,
@@ -691,6 +_,8 @@
             p_312060_,
             p_320283_
         );
+        if (configurator != null) configurator.accept(explosion); // Paper - Allow explosions to damage source
+        if (net.neoforged.neoforge.event.EventHooks.onExplosionStart(this, explosion)) return explosion;
         explosion.explode();
         explosion.finalizeExplosion(p_312145_);
         return explosion;
@@ -705,6 +_,16 @@
     @Nullable
     @Override
     public BlockEntity getBlockEntity(BlockPos p_46716_) {
+        // CraftBukkit start
+        return getBlockEntity(p_46716_, true);
+    }
+
+    @Nullable
+    public BlockEntity getBlockEntity(BlockPos p_46716_, boolean validate) {
+        if (capturedTileEntities.containsKey(p_46716_)) {
+            return capturedTileEntities.get(p_46716_);
+        }
+        // CraftBukkit end
         if (this.isOutsideBuildHeight(p_46716_)) {
             return null;
         } else {
@@ -716,7 +_,14 @@
 
     public void setBlockEntity(BlockEntity p_151524_) {
         BlockPos blockpos = p_151524_.getBlockPos();
+
         if (!this.isOutsideBuildHeight(blockpos)) {
+            // CraftBukkit start
+            if (captureBlockStates) {
+                capturedTileEntities.put(blockpos.immutable(), p_151524_);
+                return;
+            }
+            // CraftBukkit end
             this.getChunkAt(blockpos).addAndRegisterBlockEntity(p_151524_);
         }
     }
@@ -725,6 +_,7 @@
         if (!this.isOutsideBuildHeight(p_46748_)) {
             this.getChunkAt(p_46748_).removeBlockEntity(p_46748_);
         }
+        this.updateNeighbourForOutputSignal(p_46748_, getBlockState(p_46748_).getBlock()); //Notify neighbors of changes
     }
 
     public boolean isLoaded(BlockPos p_46750_) {
@@ -803,6 +_,7 @@
                 list.add(p_151522_);
             }
 
+            if (false)
             if (p_151522_ instanceof EnderDragon) {
                 for (EnderDragonPart enderdragonpart : ((EnderDragon)p_151522_).getSubEntities()) {
                     if (p_151522_ != p_46536_ && p_46538_.test(enderdragonpart)) {
@@ -811,6 +_,11 @@
                 }
             }
         });
+        for (net.neoforged.neoforge.entity.PartEntity<?> p : this.getPartEntities()) {
+            if (p != p_46536_ && p.getBoundingBox().intersects(p_46537_) && p_46538_.test(p)) {
+                list.add(p);
+            }
+        }
         return list;
     }
 
@@ -837,6 +_,8 @@
                 }
             }
 
+
+            if (false)
             if (p_261454_ instanceof EnderDragon enderdragon) {
                 for (EnderDragonPart enderdragonpart : enderdragon.getSubEntities()) {
                     T t = p_261885_.tryCast(enderdragonpart);
@@ -851,6 +_,15 @@
 
             return AbortableIterationConsumer.Continuation.CONTINUE;
         });
+        for (net.neoforged.neoforge.entity.PartEntity<?> p : this.getPartEntities()) {
+            T t = p_261885_.tryCast(p);
+            if (t != null && t.getBoundingBox().intersects(p_262086_) && p_261688_.test(t)) {
+                p_262071_.add(t);
+                if (p_262071_.size() >= p_261858_) {
+                    break;
+                }
+            }
+        }
     }
 
     @Nullable
@@ -979,16 +_,15 @@
     public abstract Scoreboard getScoreboard();
 
     public void updateNeighbourForOutputSignal(BlockPos p_46718_, Block p_46719_) {
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for(Direction direction : Direction.values()) {
             BlockPos blockpos = p_46718_.relative(direction);
             if (this.hasChunkAt(blockpos)) {
                 BlockState blockstate = this.getBlockState(blockpos);
-                if (blockstate.is(Blocks.COMPARATOR)) {
-                    this.neighborChanged(blockstate, blockpos, p_46719_, p_46718_, false);
-                } else if (blockstate.isRedstoneConductor(this, blockpos)) {
+                blockstate.onNeighborChange(this, blockpos, p_46718_);
+                if (blockstate.isRedstoneConductor(this, blockpos)) {
                     blockpos = blockpos.relative(direction);
                     blockstate = this.getBlockState(blockpos);
-                    if (blockstate.is(Blocks.COMPARATOR)) {
+                    if (blockstate.getWeakChanges(this, blockpos)) {
                         this.neighborChanged(blockstate, blockpos, p_46719_, p_46718_, false);
                     }
                 }
@@ -1078,6 +_,18 @@
         return this.biomeManager;
     }
 
+    private double maxEntityRadius = 2.0D;
+    @Override
+    public double getMaxEntityRadius() {
+        return maxEntityRadius;
+    }
+    @Override
+    public double increaseMaxEntityRadius(double value) {
+        if (value > maxEntityRadius)
+            maxEntityRadius = value;
+        return maxEntityRadius;
+    }
+
     public final boolean isDebug() {
         return this.isDebug;
     }
@@ -1105,7 +_,8 @@
         BLOCK("block"),
         MOB("mob"),
         TNT("tnt"),
-        TRIGGER("trigger");
+        TRIGGER("trigger"),
+        STANDARD("standard"); // CraftBukkit - Add STANDARD which will always use Explosion.Effect.DESTROY;
 
         public static final Codec<Level.ExplosionInteraction> CODEC = StringRepresentable.fromEnum(Level.ExplosionInteraction::values);
         private final String id;
@@ -1119,4 +_,48 @@
             return this.id;
         }
     }
+
+    // Neo: Variable day time code
+
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public abstract void setDayTimeFraction(float dayTimeFraction);
+
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public abstract float getDayTimeFraction();
+
+    /**
+     * Returns the current ratio between game ticks and clock ticks. If this value is negative, no
+     * speed has been set and those two are coupled 1:1 (i.e. vanilla mode).
+     */
+    public abstract float getDayTimePerTick();
+
+    /**
+     * DO NOT CALL.
+     * <p>
+     * Use {@link net.minecraft.server.level.ServerLevel#setDayTimePerTick(float)} instead.
+     */
+    public abstract void setDayTimePerTick(float dayTimePerTick);
+
+    // advances the fractional daytime, returns the integer part of it
+    @org.jetbrains.annotations.ApiStatus.Internal
+    protected long advanceDaytime() {
+        if (getDayTimePerTick() < 0) {
+            return 1L; // avoid doing math (and rounding errors) if no speed has been set
+        }
+        float dayTimeStep = getDayTimeFraction() + getDayTimePerTick();
+        long result = (long)dayTimeStep;
+        setDayTimeFraction(dayTimeStep - result);
+        return result;
+    }
+
+    // Paper start - respect global sound events gamerule
+    public List<net.minecraft.server.level.ServerPlayer> getPlayersForGlobalSoundGamerule() {
+        return this.getGameRules().getBoolean(GameRules.RULE_GLOBAL_SOUND_EVENTS) ? ((net.minecraft.server.level.ServerLevel) this).getServer().getPlayerList().players : ((net.minecraft.server.level.ServerLevel) this).players();
+    }
+
+    public double getGlobalSoundRangeSquared(java.util.function.Function<org.spigotmc.SpigotWorldConfig, Integer> rangeFunction) {
+        final double range = rangeFunction.apply(this.spigotConfig);
+        return range <= 0 ? 64.0 * 64.0 : range * range; // 64 is taken from default in ServerLevel#levelEvent
+    }
+    // Paper end - respect global sound events gamerule
 }
