--- a/net/minecraft/world/entity/projectile/windcharge/AbstractWindCharge.java
+++ b/net/minecraft/world/entity/projectile/windcharge/AbstractWindCharge.java
@@ -113,6 +_,7 @@
             Vec3 vec3 = Vec3.atLowerCornerOf(vec3i).multiply(0.25, 0.25, 0.25);
             Vec3 vec31 = p_325933_.getLocation().add(vec3);
             this.explode(vec31);
+            this.setRemovedReason(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
             this.discard();
         }
     }
@@ -121,6 +_,7 @@
     protected void onHit(HitResult p_326337_) {
         super.onHit(p_326337_);
         if (!this.level().isClientSide) {
+            this.setRemovedReason(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
             this.discard();
         }
     }
@@ -155,6 +_,7 @@
     public void tick() {
         if (!this.level().isClientSide && this.getBlockY() > this.level().getMaxBuildHeight() + 30) {
             this.explode(this.position());
+            this.setRemovedReason(org.bukkit.event.entity.EntityRemoveEvent.Cause.OUT_OF_WORLD); // CraftBukkit - add Bukkit remove cause
             this.discard();
         } else {
             super.tick();
