--- a/net/minecraft/world/item/trading/MerchantOffer.java
+++ b/net/minecraft/world/item/trading/MerchantOffer.java
@@ -21,7 +_,10 @@
                     Codec.INT.lenientOptionalFieldOf("specialPrice", Integer.valueOf(0)).forGetter(p_324423_ -> p_324423_.specialPriceDiff),
                     Codec.INT.lenientOptionalFieldOf("demand", Integer.valueOf(0)).forGetter(p_324040_ -> p_324040_.demand),
                     Codec.FLOAT.lenientOptionalFieldOf("priceMultiplier", Float.valueOf(0.0F)).forGetter(p_323953_ -> p_323953_.priceMultiplier),
-                    Codec.INT.lenientOptionalFieldOf("xp", Integer.valueOf(1)).forGetter(p_324202_ -> p_324202_.xp)
+                    Codec.INT.lenientOptionalFieldOf("xp", Integer.valueOf(1)).forGetter(p_324202_ -> p_324202_.xp),
+                    // Paper start
+                    Codec.BOOL.lenientOptionalFieldOf("Paper.IgnoreDiscounts", false).forGetter((merchantrecipe) -> {return merchantrecipe.ignoreDiscounts; })
+                    // Paper end
                 )
                 .apply(p_324269_, MerchantOffer::new)
     );
@@ -38,6 +_,20 @@
     public int demand;
     public float priceMultiplier;
     public int xp;
+    public boolean ignoreDiscounts; // Paper - Add ignore discounts API
+
+    // CraftBukkit start
+    private org.bukkit.craftbukkit.inventory.CraftMerchantRecipe bukkitHandle;
+    public org.bukkit.craftbukkit.inventory.CraftMerchantRecipe asBukkit() {
+        return (bukkitHandle == null) ? bukkitHandle = new org.bukkit.craftbukkit.inventory.CraftMerchantRecipe(this) : bukkitHandle;
+    }
+
+    public MerchantOffer(ItemCost baseCostA, Optional<ItemCost> costB, ItemStack result, int uses, int maxUses, int experience, float priceMultiplier, int demand, final boolean ignoreDiscounts, org.bukkit.craftbukkit.inventory.CraftMerchantRecipe bukkit) { // Paper
+        this(baseCostA, costB, result, uses, maxUses, experience, priceMultiplier, demand);
+        this.ignoreDiscounts = ignoreDiscounts; // Paper
+        this.bukkitHandle = bukkit;
+    }
+    // CraftBukkit end
 
     private MerchantOffer(
         ItemCost p_330734_,
@@ -49,7 +_,8 @@
         int p_332006_,
         int p_330432_,
         float p_45332_,
-        int p_330282_
+        int p_330282_,
+        final boolean ignoreDiscounts // Paper
     ) {
         this.baseCostA = p_330734_;
         this.costB = p_331500_;
@@ -61,6 +_,7 @@
         this.demand = p_330432_;
         this.priceMultiplier = p_45332_;
         this.xp = p_330282_;
+        this.ignoreDiscounts = ignoreDiscounts; // Paper
     }
 
     public MerchantOffer(ItemCost p_332077_, ItemStack p_320418_, int p_320071_, int p_320069_, float p_320947_) {
@@ -78,7 +_,7 @@
     public MerchantOffer(
         ItemCost p_331744_, Optional<ItemCost> p_330460_, ItemStack p_324239_, int p_324562_, int p_324493_, int p_323558_, float p_323528_, int p_324484_
     ) {
-        this(p_331744_, p_330460_, p_324239_, p_324562_, p_324493_, true, 0, p_324484_, p_323528_, p_323558_);
+        this(p_331744_, p_330460_, p_324239_, p_324562_, p_324493_, true, 0, p_324484_, p_323528_, p_323558_, false); // Paper
     }
 
     private MerchantOffer(MerchantOffer p_302340_) {
@@ -92,7 +_,8 @@
             p_302340_.specialPriceDiff,
             p_302340_.demand,
             p_302340_.priceMultiplier,
-            p_302340_.xp
+            p_302340_.xp,
+            p_302340_.ignoreDiscounts // Paper
         );
     }
 
@@ -127,7 +_,7 @@
     }
 
     public void updateDemand() {
-        this.demand = this.demand + this.uses - (this.maxUses - this.uses);
+        this.demand = Math.max(0, this.demand + this.uses - (this.maxUses - this.uses)); // Paper - Fix MC-163962
     }
 
     public ItemStack assemble() {
@@ -206,7 +_,11 @@
         if (!this.satisfiedBy(p_45362_, p_45363_)) {
             return false;
         } else {
-            p_45362_.shrink(this.getCostA().getCount());
+            // CraftBukkit start
+            if (!this.getCostA().isEmpty()) {
+                p_45363_.shrink(this.getCostA().getCount());
+            }
+            // CraftBukkit end
             if (!this.getCostB().isEmpty()) {
                 p_45363_.shrink(this.getCostB().getCount());
             }
