--- a/net/minecraft/world/entity/ai/sensing/Sensor.java
+++ b/net/minecraft/world/entity/ai/sensing/Sensor.java
@@ -26,8 +_,21 @@
         .ignoreInvisibilityTesting();
     private final int scanRate;
     private long timeToTick;
+    // Paper start - configurable sensor tick rate and timings
+    private final String configKey;
+    private final co.aikar.timings.Timing timing;
+    // Paper end
 
     public Sensor(int p_26800_) {
+        // Paper start - configurable sensor tick rate and timings
+        String key = io.papermc.paper.util.MappingEnvironment.reobf() ? io.papermc.paper.util.ObfHelper.INSTANCE.deobfClassName(this.getClass().getName()) : this.getClass().getName();
+        int lastSeparator = key.lastIndexOf('.');
+        if (lastSeparator != -1) {
+            key = key.substring(lastSeparator + 1);
+        }
+        this.configKey = key.toLowerCase(java.util.Locale.ROOT);
+        this.timing = co.aikar.timings.MinecraftTimings.getSensorTimings(configKey, p_26800_);
+        // Paper end
         this.scanRate = p_26800_;
         this.timeToTick = (long)RANDOM.nextInt(p_26800_);
     }
@@ -38,8 +_,12 @@
 
     public final void tick(ServerLevel p_26807_, E p_26808_) {
         if (--this.timeToTick <= 0L) {
-            this.timeToTick = (long)this.scanRate;
+            // Paper start - configurable sensor tick rate and timings
+            this.timeToTick = java.util.Objects.requireNonNullElse(p_26807_.paperConfig().tickRates.sensor.get(p_26808_.getType(), this.configKey), this.scanRate);
+            this.timing.startTiming();
+            // Paper end
             this.doTick(p_26807_, p_26808_);
+            this.timing.stopTiming(); // Paper - sensor timings
         }
     }
 
