--- a/net/minecraft/world/entity/TamableAnimal.java
+++ b/net/minecraft/world/entity/TamableAnimal.java
@@ -91,8 +_,12 @@
     @Override
     public boolean handleLeashAtDistance(Entity p_352258_, float p_352278_) {
         if (this.isInSittingPose()) {
-            if (p_352278_ > 10.0F) {
-                this.dropLeash(true, true);
+            if (p_352278_ > (float) this.level().paperConfig().misc.maxLeashDistance.or(Leashable.LEASH_TOO_FAR_DIST)) { // Paper - Configurable max leash distance
+                // Paper start - Expand EntityUnleashEvent
+                org.bukkit.event.entity.EntityUnleashEvent event = new org.bukkit.event.entity.EntityUnleashEvent(this.getBukkitEntity(), org.bukkit.event.entity.EntityUnleashEvent.UnleashReason.DISTANCE, true);
+                if (!event.callEvent()) return false;
+                this.dropLeash(true, event.isDropLeash());
+                // Paper end - Expand EntityUnleashEvent
             }
 
             return false;
@@ -220,11 +_,20 @@
 
     @Override
     public void die(DamageSource p_21809_) {
-        if (!this.level().isClientSide && this.level().getGameRules().getBoolean(GameRules.RULE_SHOWDEATHMESSAGES) && this.getOwner() instanceof ServerPlayer) {
-            this.getOwner().sendSystemMessage(this.getCombatTracker().getDeathMessage());
-        }
-
+        // FORGE: Super moved to top so that death message would be cancelled properly
+        net.minecraft.network.chat.Component deathMessage = this.getCombatTracker().getDeathMessage();
         super.die(p_21809_);
+
+        if (this.dead)
+        if (!this.level().isClientSide && this.level().getGameRules().getBoolean(GameRules.RULE_SHOWDEATHMESSAGES) && this.getOwner() instanceof ServerPlayer) {
+            // Paper start - Add TameableDeathMessageEvent
+            io.papermc.paper.event.entity.TameableDeathMessageEvent event = new io.papermc.paper.event.entity.TameableDeathMessageEvent((org.bukkit.entity.Tameable) getBukkitEntity(), io.papermc.paper.adventure.PaperAdventure.asAdventure(this.getCombatTracker().getDeathMessage()));
+            if (event.callEvent()) {
+                this.getOwner().sendSystemMessage(io.papermc.paper.adventure.PaperAdventure.asVanilla(event.deathMessage()));
+            }
+            // Paper end - Add TameableDeathMessageEvent
+        }
+
     }
 
     public boolean isOrderedToSit() {
@@ -264,7 +_,14 @@
         if (!this.canTeleportTo(new BlockPos(p_350930_, p_350303_, p_350410_))) {
             return false;
         } else {
-            this.moveTo((double)p_350930_ + 0.5, (double)p_350303_, (double)p_350410_ + 0.5, this.getYRot(), this.getXRot());
+            // CraftBukkit start
+            org.bukkit.event.entity.EntityTeleportEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTeleportEvent(this, (double) p_350930_ + 0.5D, (double) p_350303_, (double) p_350410_ + 0.5D);
+            if (event.isCancelled()) {
+                return false;
+            }
+            org.bukkit.Location to = event.getTo();
+            this.moveTo(to.getX(), to.getY(), to.getZ(), to.getYaw(), to.getPitch());
+            // CraftBukkit end
             this.navigation.stop();
             return true;
         }
