--- a/net/minecraft/gametest/framework/GameTestServer.java
+++ b/net/minecraft/gametest/framework/GameTestServer.java
@@ -151,13 +_,15 @@
     public boolean initServer() {
         this.setPlayerList(new PlayerList(this, this.registries(), this.playerDataStorage, 1) {
         });
-        this.loadLevel();
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerAboutToStart(this);
+        this.loadLevel(storageSource.getLevelId()); // CraftBukkit
         ServerLevel serverlevel = this.overworld();
         this.testBatches = Lists.newArrayList(GameTestBatchFactory.fromTestFunction(this.testFunctions, serverlevel));
         serverlevel.setDefaultSpawnPos(this.spawnPos, 0.0F);
         int i = 20000000;
         serverlevel.setWeatherParameters(20000000, 20000000, false, false);
         LOGGER.info("Started game test server");
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStarting(this);
         return true;
     }
 
@@ -296,6 +_,12 @@
     @Override
     public boolean shouldInformAdmins() {
         return false;
+    }
+
+    @Override
+    public org.bukkit.command.CommandSender getBukkitSender(net.minecraft.commands.CommandSourceStack wrapper) {
+
+        return null;
     }
 
     @Override
