--- a/net/minecraft/world/entity/projectile/ThrownTrident.java
+++ b/net/minecraft/world/entity/projectile/ThrownTrident.java
@@ -67,6 +_,7 @@
                     this.spawnAtLocation(this.getPickupItem(), 0.1F);
                 }
 
+                this.setRemovedReason(org.bukkit.event.entity.EntityRemoveEvent.Cause.DROP); // CraftBukkit - add Bukkit remove cause
                 this.discard();
             } else {
                 this.setNoPhysics(true);
