--- a/net/minecraft/world/entity/projectile/ThrownTrident.java
+++ b/net/minecraft/world/entity/projectile/ThrownTrident.java
@@ -67,6 +_,7 @@
                     this.spawnAtLocation(this.getPickupItem(), 0.1F);
                 }
 
+                this.setRemovedReason(org.bukkit.event.entity.EntityRemoveEvent.Cause.DROP); // CraftBukkit - add Bukkit remove cause
                 this.discard();
             } else {
                 this.setNoPhysics(true);
@@ -97,6 +_,20 @@
     public boolean isFoil() {
         return this.entityData.get(ID_FOIL);
     }
+
+    // Paper start
+    public void setFoil(boolean foil) {
+        this.entityData.set(ThrownTrident.ID_FOIL, foil);
+    }
+
+    public int getLoyalty() {
+        return this.entityData.get(ThrownTrident.ID_LOYALTY);
+    }
+
+    public void setLoyalty(byte loyalty) {
+        this.entityData.set(ThrownTrident.ID_LOYALTY, loyalty);
+    }
+    // Paper end
 
     @Nullable
     @Override
