--- a/net/minecraft/world/entity/npc/VillagerTrades.java
+++ b/net/minecraft/world/entity/npc/VillagerTrades.java
@@ -1477,6 +_,7 @@
         private final int villagerXp;
 
         public EmeraldsForVillagerTypeItem(int p_35669_, int p_35670_, int p_35671_, Map<VillagerType, Item> p_35672_) {
+            if (false) // Neo: disable this check so that mods can add custom villager types
             BuiltInRegistries.VILLAGER_TYPE.stream().filter(p_35680_ -> !p_35672_.containsKey(p_35680_)).findAny().ifPresent(p_339515_ -> {
                 throw new IllegalStateException("Missing trade for villager type: " + BuiltInRegistries.VILLAGER_TYPE.getKey(p_339515_));
             });
@@ -1490,7 +_,9 @@
         @Override
         public MerchantOffer getOffer(Entity p_219685_, RandomSource p_219686_) {
             if (p_219685_ instanceof VillagerDataHolder villagerdataholder) {
-                ItemCost itemcost = new ItemCost(this.trades.get(villagerdataholder.getVillagerData().getType()), this.cost);
+                Item item = this.trades.get(villagerdataholder.getVillagerData().getType());
+                if (item == null) return null;  // Neo: add a check for unknown villager types
+                ItemCost itemcost = new ItemCost(item, this.cost);
                 return new MerchantOffer(itemcost, new ItemStack(Items.EMERALD), this.maxUses, this.villagerXp, 0.05F);
             } else {
                 return null;
@@ -1825,7 +_,8 @@
                 return null;
             } else {
                 ServerLevel serverlevel = (ServerLevel)p_219708_.level();
-                BlockPos blockpos = serverlevel.findNearestMapStructure(this.destination, p_219708_.blockPosition(), 100, true);
+                if (!serverlevel.paperConfig().environment.treasureMaps.enabled) return null; // Paper - Configurable cartographer treasure maps
+                BlockPos blockpos = serverlevel.findNearestMapStructure(this.destination, p_219708_.blockPosition(), 100, !serverlevel.paperConfig().environment.treasureMaps.findAlreadyDiscoveredVillager); // Paper - Configurable cartographer treasure maps
                 if (blockpos != null) {
                     ItemStack itemstack = MapItem.create(serverlevel, blockpos.getX(), blockpos.getZ(), (byte)2, true, true);
                     MapItem.renderBiomePreviewMap(serverlevel, itemstack);
