--- a/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -115,6 +_,12 @@
         if (p_289768_.isRespawning) {
             this.respawnStage = DragonRespawnAnimation.START;
         }
+        // Paper start - Add config to disable ender dragon legacy check
+        if (p_289768_ == EndDragonFight.Data.DEFAULT && !p_289771_.paperConfig().entities.spawning.scanForLegacyEnderDragon) {
+            this.needsStateScanning = false;
+            this.dragonKilled = true;
+        }
+        // Paper end - Add config to disable ender dragon legacy check
 
         this.portalLocation = p_289768_.exitPortalLocation.orElse(null);
         this.gateways.addAll(p_289768_.gateways.orElseGet(() -> {
@@ -212,8 +_,9 @@
             this.dragonUUID = enderdragon.getUUID();
             LOGGER.info("Found that there's a dragon still alive ({})", enderdragon);
             this.dragonKilled = false;
-            if (!flag) {
+            if (!flag && this.level.paperConfig().entities.behavior.shouldRemoveDragon) { // Paper - Toggle for removing existing dragon
                 LOGGER.info("But we didn't have a portal, let's remove it.");
+                enderdragon.setRemovedReason(null);
                 enderdragon.discard();
                 this.dragonUUID = null;
             }
@@ -370,12 +_,21 @@
             this.dragonEvent.setVisible(false);
             this.spawnExitPortal(true);
             this.spawnNewGateway();
-            if (!this.previouslyKilled) {
-                this.level
-                    .setBlockAndUpdate(
-                        this.level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, EndPodiumFeature.getLocation(this.origin)),
-                        Blocks.DRAGON_EGG.defaultBlockState()
-                    );
+            // Paper start - Add DragonEggFormEvent
+            BlockPos eggPosition = this.level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, EndPodiumFeature.getLocation(this.origin));
+            org.bukkit.craftbukkit.block.CraftBlockState eggState = org.bukkit.craftbukkit.block.CraftBlockStates.getBlockState(this.level, eggPosition);
+            eggState.setData(Blocks.DRAGON_EGG.defaultBlockState());
+            io.papermc.paper.event.block.DragonEggFormEvent eggEvent = new io.papermc.paper.event.block.DragonEggFormEvent(org.bukkit.craftbukkit.block.CraftBlock.at(this.level, eggPosition), eggState,
+                    new org.bukkit.craftbukkit.boss.CraftDragonBattle(this));
+            // Paper end - Add DragonEggFormEvent
+            if (this.level.paperConfig().entities.behavior.enderDragonsDeathAlwaysPlacesDragonEgg || !this.previouslyKilled) { // Paper - Add toggle for always placing the dragon egg
+                // Paper start - Add DragonEggFormEvent
+            } else {
+                eggEvent.setCancelled(true);
+            }
+            if (eggEvent.callEvent()) {
+                eggEvent.getNewState().update(true);
+                // Paper end - Add DragonEggFormEvent
             }
 
             this.previouslyKilled = true;
@@ -417,6 +_,11 @@
             }
         }
 
+        // Paper start - Prevent "softlocked" exit portal generation
+        if (this.portalLocation.getY() <= this.level.getMinBuildHeight()) {
+            this.portalLocation = this.portalLocation.atY(this.level.getMinBuildHeight() + 1);
+        }
+        // Paper end - Prevent "softlocked" exit portal generation
         if (endpodiumfeature.place(
             FeatureConfiguration.NONE, this.level, this.level.getChunkSource().getGenerator(), RandomSource.create(), this.portalLocation
         )) {
@@ -438,6 +_,7 @@
             );
             this.level.addFreshEntity(enderdragon);
             this.dragonUUID = enderdragon.getUUID();
+            this.resetSpikeCrystals(); // Paper - Reset ender crystals on dragon spawn
         }
 
         return enderdragon;
@@ -477,6 +_,13 @@
         return this.previouslyKilled;
     }
 
+    // Magma
+    private java.util.concurrent.atomic.AtomicBoolean magma$tryRespawn = new java.util.concurrent.atomic.AtomicBoolean(false);
+    private java.util.concurrent.atomic.AtomicBoolean magma$respawnDragon = new java.util.concurrent.atomic.AtomicBoolean(false);
+
+    public boolean magma$tryRespawn() { return magma$tryRespawn.getAndSet(false); }
+    public boolean magma$respawnDragon() { return magma$respawnDragon.getAndSet(false); }
+
     public void tryRespawn() {
         if (this.dragonKilled && this.respawnStage == null) {
             BlockPos blockpos = this.portalLocation;
@@ -499,15 +_,16 @@
             for (Direction direction : Direction.Plane.HORIZONTAL) {
                 List<EndCrystal> list = this.level.getEntitiesOfClass(EndCrystal.class, new AABB(blockpos1.relative(direction, 2)));
                 if (list.isEmpty()) {
-                    return;
+                    magma$tryRespawn.set(true);
                 }
 
                 list1.addAll(list);
             }
 
             LOGGER.debug("Found all crystals, respawning dragon.");
-            this.respawnDragon(list1);
+            magma$tryRespawn.set(magma$respawnDragon());
         }
+        magma$tryRespawn.set(false);
     }
 
     public void respawnDragon(List<EndCrystal> p_64092_) {
@@ -532,7 +_,9 @@
             this.respawnTime = 0;
             this.spawnExitPortal(false);
             this.respawnCrystals = p_64092_;
+            magma$respawnDragon.set(true);
         }
+        magma$respawnDragon.set(false);
     }
 
     public void resetSpikeCrystals() {
@@ -542,6 +_,14 @@
                 endcrystal.setBeamTarget(null);
             }
         }
+    }
+
+    public void addPlayer(ServerPlayer player) {
+        this.dragonEvent.addPlayer(player);
+    }
+
+    public void removePlayer(ServerPlayer player) {
+        this.dragonEvent.removePlayer(player);
     }
 
     @Nullable
