--- a/net/minecraft/world/level/block/BedBlock.java
+++ b/net/minecraft/world/level/block/BedBlock.java
@@ -78,6 +_,10 @@
         return blockstate.getBlock() instanceof BedBlock ? blockstate.getValue(FACING) : null;
     }
 
+    private BlockState finalBlockState;
+    private BlockPos finalBlockPos;
+    private Level finalLevel;
+
     @Override
     protected InteractionResult useWithoutItem(BlockState p_49515_, Level p_49516_, BlockPos p_49517_, Player p_49518_, BlockHitResult p_49520_) {
         if (p_49516_.isClientSide) {
@@ -91,7 +_,8 @@
                 }
             }
 
-            if (!canSetSpawn(p_49516_)) {
+            // CraftBukkit - moved world and biome check into EntityHuman
+            if (false && !canSetSpawn(p_49516_)) {
                 p_49516_.removeBlock(p_49517_, false);
                 BlockPos blockpos = p_49517_.relative(p_49515_.getValue(FACING).getOpposite());
                 if (p_49516_.getBlockState(blockpos).is(this)) {
@@ -108,18 +_,57 @@
 
                 return InteractionResult.SUCCESS;
             } else {
+                finalBlockPos = p_49517_;
+                finalBlockState = p_49515_;
+                finalLevel = p_49516_;
                 p_49518_.startSleepInBed(p_49517_).ifLeft(p_49477_ -> {
-                    if (p_49477_.getMessage() != null) {
-                        p_49518_.displayClientMessage(p_49477_.getMessage(), true);
-                    }
+                    // Paper start - PlayerBedFailEnterEvent
+                    if (p_49477_ != null) {
+                        io.papermc.paper.event.player.PlayerBedFailEnterEvent event = new io.papermc.paper.event.player.PlayerBedFailEnterEvent((org.bukkit.entity.Player) p_49518_.getBukkitEntity(), io.papermc.paper.event.player.PlayerBedFailEnterEvent.FailReason.values()[p_49477_.ordinal()], org.bukkit.craftbukkit.block.CraftBlock.at(p_49516_, finalBlockPos), !p_49516_.dimensionType().bedWorks(), io.papermc.paper.adventure.PaperAdventure.asAdventure(p_49477_.getMessage()));
+                        if (!event.callEvent()) {
+                            return;
+                        }
+                        // Paper end - PlayerBedFailEnterEvent
+                        // CraftBukkit start - handling bed explosion from below here
+                        if (event.getWillExplode()) { // Paper - PlayerBedFailEnterEvent
+                            this.explodeBed(finalBlockState, finalLevel, finalBlockPos);
+                        } else
+                            // CraftBukkit end
+                            if (p_49477_.getMessage() != null) {
+                                final net.kyori.adventure.text.Component message = event.getMessage(); // Paper - PlayerBedFailEnterEvent
+                                if (message != null) p_49518_.displayClientMessage(io.papermc.paper.adventure.PaperAdventure.asVanilla(message), true); // Paper - PlayerBedFailEnterEvent
+                            }
+                    } // Paper - PlayerBedFailEnterEvent
                 });
                 return InteractionResult.SUCCESS;
             }
         }
     }
 
+    // CraftBukkit start
+    private InteractionResult explodeBed(BlockState blockState, Level world, BlockPos blockPos) {
+        {
+            {
+                org.bukkit.block.BlockState blockState2 = org.bukkit.craftbukkit.block.CraftBlock.at(world, blockPos).getState(); // CraftBukkit - capture BlockState before remove block
+                world.removeBlock(blockPos, false);
+                BlockPos blockPos1 = blockPos.relative(blockState.getValue(FACING).getOpposite());
+
+                if (world.getBlockState(blockPos1).getBlock() == this) {
+                    world.removeBlock(blockPos1, false);
+                }
+
+                Vec3 vec3 = blockPos.getCenter();
+
+                world.explode(null, world.damageSources().badRespawnPointExplosion(vec3, blockState2), null, vec3, 5.0F, true, Level.ExplosionInteraction.BLOCK); // CraftBukkit - add state
+                return InteractionResult.SUCCESS;
+            }
+        }
+    }
+    // CraftBukkit end
+
     public static boolean canSetSpawn(Level p_49489_) {
-        return p_49489_.dimensionType().bedWorks();
+        // CraftBukkit - moved world and biome check into EntityHuman
+        return true || p_49489_.dimensionType().bedWorks();
     }
 
     private boolean kickVillagerOutOfBed(Level p_49491_, BlockPos p_49492_) {
@@ -308,6 +_,11 @@
         if (!p_49499_.isClientSide) {
             BlockPos blockpos = p_49500_.relative(p_49501_.getValue(FACING));
             p_49499_.setBlock(blockpos, p_49501_.setValue(PART, BedPart.HEAD), 3);
+            // CraftBukkit start - SPIGOT-7315: Don't updated if we capture block states
+            if (p_49499_.captureBlockStates) {
+                return;
+            }
+            // CraftBukkit end
             p_49499_.blockUpdated(p_49500_, Blocks.AIR);
             p_49501_.updateNeighbourShapes(p_49499_, p_49500_, 3);
         }
