--- a/net/minecraft/world/level/block/NetherPortalBlock.java
+++ b/net/minecraft/world/level/block/NetherPortalBlock.java
@@ -38,6 +_,7 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 import org.slf4j.Logger;
 
+// todo: Magma - Add CB Events here with mixin protections
 public class NetherPortalBlock extends Block implements Portal {
     public static final MapCodec<NetherPortalBlock> CODEC = simpleCodec(NetherPortalBlock::new);
     public static final EnumProperty<Direction.Axis> AXIS = BlockStateProperties.HORIZONTAL_AXIS;
@@ -69,7 +_,7 @@
 
     @Override
     protected void randomTick(BlockState p_221799_, ServerLevel p_221800_, BlockPos p_221801_, RandomSource p_221802_) {
-        if (p_221800_.dimensionType().natural()
+        if (p_221800_.spigotConfig.enableZombiePigmenPortalSpawns && p_221800_.dimensionType().natural() // Spigot
             && p_221800_.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)
             && p_221802_.nextInt(2000) < p_221800_.getDifficulty().getId()) {
             while (p_221800_.getBlockState(p_221801_).is(this)) {
@@ -80,6 +_,8 @@
                 Entity entity = EntityType.ZOMBIFIED_PIGLIN.spawn(p_221800_, p_221801_.above(), MobSpawnType.STRUCTURE);
                 if (entity != null) {
                     entity.setPortalCooldown();
+                    entity.fromNetherPortal = true; // Paper - Add option to nerf pigmen from nether portals
+                    if (p_221800_.paperConfig().entities.behavior.nerfPigmenFromNetherPortals) ((net.minecraft.world.entity.Mob) entity).aware = false; // Paper - Add option to nerf pigmen from nether portals
                 }
             }
         }
@@ -129,6 +_,13 @@
             WorldBorder worldborder = serverlevel.getWorldBorder();
             double d0 = DimensionType.getTeleportationScale(p_350444_.dimensionType(), serverlevel.dimensionType());
             BlockPos blockpos = worldborder.clampToBounds(p_350334_.getX() * d0, p_350334_.getY(), p_350334_.getZ() * d0);
+            // Paper start - Configurable portal search radius
+            int portalSearchRadius = serverlevel.paperConfig().environment.portalSearchRadius;
+            if (p_350334_.level().paperConfig().environment.portalSearchVanillaDimensionScaling && flag) { // flag = is going to nether
+                portalSearchRadius = (int) (portalSearchRadius / serverlevel.dimensionType().coordinateScale());
+            }
+            // Paper end - Configurable portal search radius
+            // magma todo - // CraftPortalEvent event = entity.callPortalEvent(entity, CraftLocation.toBukkit(blockposition1, worldserver1.getWorld()), PlayerTeleportEvent.TeleportCause.NETHER_PORTAL, portalSearchRadius, worldserver1.paperConfig().environment.portalCreateRadius); // Paper - use custom portal search radius
             return this.getExitPortal(serverlevel, p_350334_, p_350764_, blockpos, flag, worldborder);
         }
     }
