--- a/net/minecraft/world/level/block/SignBlock.java
+++ b/net/minecraft/world/level/block/SignBlock.java
@@ -127,6 +_,7 @@
             } else if (!this.otherPlayerIsEditingSign(p_316173_, signblockentity)
                 && p_316173_.mayBuild()
                 && this.hasEditableText(p_316173_, signblockentity, flag1)) {
+                this.setOpenSignCause(org.bukkit.event.player.PlayerSignOpenEvent.Cause.INTERACT);
                 this.openTextEdit(p_316173_, signblockentity, flag1);
                 return InteractionResult.SUCCESS;
             } else {
@@ -169,7 +_,19 @@
         return woodtype;
     }
 
+    // Magma start - mixin support
+    private org.bukkit.event.player.PlayerSignOpenEvent.Cause magma$PlayerSignOpenCause = org.bukkit.event.player.PlayerSignOpenEvent.Cause.UNKNOWN;
+    public void setOpenSignCause(org.bukkit.event.player.PlayerSignOpenEvent.Cause cause) {
+        this.magma$PlayerSignOpenCause = cause;
+    }
+
     public void openTextEdit(Player p_277738_, SignBlockEntity p_277467_, boolean p_277771_) {
+        var cause = magma$PlayerSignOpenCause;
+        if (!org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerSignOpenEvent(p_277738_, p_277467_, p_277771_, cause)) {
+            return;
+        }
+        magma$PlayerSignOpenCause = org.bukkit.event.player.PlayerSignOpenEvent.Cause.UNKNOWN;
+        // CraftBukkit end
         p_277467_.setAllowedPlayerEditor(p_277738_.getUUID());
         p_277738_.openTextEdit(p_277467_, p_277771_);
     }
@@ -182,6 +_,6 @@
     @Nullable
     @Override
     public <T extends BlockEntity> BlockEntityTicker<T> getTicker(Level p_277367_, BlockState p_277896_, BlockEntityType<T> p_277724_) {
-        return createTickerHelper(p_277724_, BlockEntityType.SIGN, SignBlockEntity::tick);
+        return null; // CraftBukkit - remove unnecessary sign ticking
     }
 }
