--- a/net/minecraft/world/entity/ai/navigation/WallClimberNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/WallClimberNavigation.java
@@ -16,9 +_,9 @@
     }
 
     @Override
-    public Path createPath(BlockPos p_26589_, int p_26590_) {
+    public Path createPath(BlockPos p_26589_, @Nullable Entity entity, int p_26590_) { // Paper - EntityPathfindEvent
         this.pathToPosition = p_26589_;
-        return super.createPath(p_26589_, p_26590_);
+        return super.createPath(p_26589_, entity, p_26590_); // Paper - EntityPathfindEvent
     }
 
     @Override
@@ -45,11 +_,12 @@
             super.tick();
         } else {
             if (this.pathToPosition != null) {
-                if (!this.pathToPosition.closerToCenterThan(this.mob.position(), (double)this.mob.getBbWidth())
+                // FORGE: Fix MC-94054
+                if (!this.pathToPosition.closerToCenterThan(this.mob.position(), Math.max((double)this.mob.getBbWidth(), 1.0D))
                     && (
                         !(this.mob.getY() > (double)this.pathToPosition.getY())
                             || !BlockPos.containing((double)this.pathToPosition.getX(), this.mob.getY(), (double)this.pathToPosition.getZ())
-                                .closerToCenterThan(this.mob.position(), (double)this.mob.getBbWidth())
+                                .closerToCenterThan(this.mob.position(), Math.max((double)this.mob.getBbWidth(), 1.0D))
                     )) {
                     this.mob
                         .getMoveControl()
