--- a/net/minecraft/world/Container.java
+++ b/net/minecraft/world/Container.java
@@ -91,4 +_,40 @@
             return level.getBlockEntity(blockpos) != p_272877_ ? false : p_272670_.canInteractWithBlock(blockpos, (double)p_320837_);
         }
     }
+
+    // CraftBukkit start
+    default java.util.List<ItemStack> getContents() {
+        return java.util.Collections.emptyList();
+    }
+
+    default void onOpen(org.bukkit.craftbukkit.entity.CraftHumanEntity who) {
+
+    }
+
+    default void onClose(org.bukkit.craftbukkit.entity.CraftHumanEntity who) {
+
+    }
+
+    default java.util.List<org.bukkit.entity.HumanEntity> getViewers() {
+        return java.util.Collections.emptyList();
+    }
+
+    default org.bukkit.inventory.InventoryHolder getOwner() {
+        return this instanceof BlockEntity blockEntity ? org.magmafoundation.magma.util.InventoryHelper.getHolder(blockEntity) : null;
+    }
+
+    default void setMaxStackSize(int size) {
+
+    }
+
+    default org.bukkit.Location getLocation() {
+        if(this instanceof BlockEntity blockEntity) {
+            BlockPos pos = blockEntity.getBlockPos();
+            return new org.bukkit.Location(blockEntity.getLevel().getWorld(), pos.getX(), pos.getY(), pos.getZ());
+        }
+        return null;
+    }
+
+    int MAX_STACK = 99;
+    // CraftBukkit end
 }
