--- a/net/minecraft/world/entity/raid/Raider.java
+++ b/net/minecraft/world/entity/raid/Raider.java
@@ -222,6 +_,7 @@
             this.onItemPickup(p_37866_);
             this.setItemSlot(equipmentslot, itemstack);
             this.take(p_37866_, itemstack.getCount());
+            p_37866_.setRemovedReason(org.bukkit.event.entity.EntityRemoveEvent.Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
             p_37866_.discard();
             this.getCurrentRaid().setLeader(this.getWave(), this);
             this.setPatrolLeader(true);
@@ -279,6 +_,7 @@
 
         @Override
         public boolean canUse() {
+            if (!this.mob.level().getGameRules().getBoolean(net.minecraft.world.level.GameRules.RULE_MOBGRIEFING) || !this.mob.canPickUpLoot()) return false; // Paper - respect game and entity rules for picking up items
             LivingEntity livingentity = this.mob.getLastHurtByMob();
             return this.mob.getCurrentRaid() == null
                 && this.mob.isPatrolling()
@@ -295,7 +_,7 @@
             for (Raider raider : this.mob
                 .level()
                 .getNearbyEntities(Raider.class, this.shoutTargeting, this.mob, this.mob.getBoundingBox().inflate(8.0, 8.0, 8.0))) {
-                raider.setTarget(this.mob.getTarget());
+                raider.setTarget(this.mob.getTarget(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER, true); // CraftBukkit
             }
         }
 
@@ -307,7 +_,7 @@
                 for (Raider raider : this.mob
                     .level()
                     .getNearbyEntities(Raider.class, this.shoutTargeting, this.mob, this.mob.getBoundingBox().inflate(8.0, 8.0, 8.0))) {
-                    raider.setTarget(livingentity);
+                    raider.setTarget(livingentity, org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER, true); // CraftBukkit
                     raider.setAggressive(true);
                 }
 
