--- a/net/minecraft/world/entity/decoration/ItemFrame.java
+++ b/net/minecraft/world/entity/decoration/ItemFrame.java
@@ -87,6 +_,12 @@
 
     @Override
     protected AABB calculateBoundingBox(BlockPos p_345363_, Direction p_344892_) {
+        // CraftBukkit start - break out BB calc into own method
+        return calculateBoundingBoxStatic(p_345363_, p_344892_);
+    }
+
+    public static AABB calculateBoundingBoxStatic(BlockPos p_345363_, Direction p_344892_) {
+        // CraftBukkit end
         float f = 0.46875F;
         Vec3 vec3 = Vec3.atCenterOf(p_345363_).relative(p_344892_, -0.46875);
         Direction.Axis direction$axis = p_344892_.getAxis();
@@ -138,6 +_,11 @@
             return false;
         } else if (!p_31776_.is(DamageTypeTags.IS_EXPLOSION) && !this.getItem().isEmpty()) {
             if (!this.level().isClientSide) {
+                // CraftBukkit start - fire EntityDamageEvent
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_31776_, p_31777_, false) || this.isRemoved()) {
+                    return true;
+                }
+                // CraftBukkit end
                 this.dropItem(p_31776_.getEntity(), false);
                 this.gameEvent(GameEvent.BLOCK_CHANGE, p_31776_.getEntity());
                 this.playSound(this.getRemoveItemSound(), 1.0F, 1.0F);
@@ -240,13 +_,19 @@
     }
 
     public void setItem(ItemStack p_31790_, boolean p_31791_) {
+        // CraftBukkit start
+        this.setItem(p_31790_, p_31791_, true);
+    }
+
+    public void setItem(ItemStack p_31790_, boolean p_31791_, boolean playSound) {
+        // CraftBukkit end
         if (!p_31790_.isEmpty()) {
             p_31790_ = p_31790_.copyWithCount(1);
         }
 
         this.onItemChanged(p_31790_);
         this.getEntityData().set(DATA_ITEM, p_31790_);
-        if (!p_31790_.isEmpty()) {
+        if (!p_31790_.isEmpty() && playSound) { // CraftBukkit
             this.playSound(this.getAddItemSound(), 1.0F, 1.0F);
         }
 
