--- a/net/minecraft/world/level/chunk/storage/ChunkStorage.java
+++ b/net/minecraft/world/level/chunk/storage/ChunkStorage.java
@@ -38,17 +_,63 @@
         return this.worker.isOldChunkAround(p_223452_, p_223453_);
     }
 
+    // CraftBukkit start
+    private boolean check(net.minecraft.server.level.ServerChunkCache cps, int x, int z) {
+        ChunkPos pos = new ChunkPos(x, z);
+        if (cps != null) {
+            com.google.common.base.Preconditions.checkState(org.bukkit.Bukkit.isPrimaryThread(), "primary thread");
+            if (cps.hasChunk(x, z)) {
+                return true;
+            }
+        }
+
+        CompoundTag nbt;
+        try {
+            nbt = read(pos).get().orElse(null);
+        } catch (InterruptedException | java.util.concurrent.ExecutionException ex) {
+            throw new RuntimeException(ex);
+        }
+        if (nbt != null) {
+            CompoundTag level = nbt.getCompound("Level");
+            if (level.getBoolean("TerrainPopulated")) {
+                return true;
+            }
+
+            net.minecraft.world.level.chunk.status.ChunkStatus status = net.minecraft.world.level.chunk.status.ChunkStatus.byName(level.getString("Status"));
+            if (status != null && status.isOrAfter(net.minecraft.world.level.chunk.status.ChunkStatus.FEATURES)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+    // CraftBukkit end
+
     public CompoundTag upgradeChunkTag(
         ResourceKey<Level> p_188289_,
         Supplier<DimensionDataStorage> p_188290_,
         CompoundTag p_188291_,
-        Optional<ResourceKey<MapCodec<? extends ChunkGenerator>>> p_188292_
+        Optional<ResourceKey<MapCodec<? extends ChunkGenerator>>> p_188292_,
+        ChunkPos pos, // CraftBukkit
+        @Nullable net.minecraft.server.level.ServerLevel generatoraccess // CraftBukkit
     ) {
         int i = getVersion(p_188291_);
         if (i == SharedConstants.getCurrentVersion().getDataVersion().getVersion()) {
             return p_188291_;
         } else {
             try {
+                // CraftBukkit start
+                if (i < 1466) {
+                    CompoundTag level = p_188291_.getCompound("Level");
+                    if (level.getBoolean("TerrainPopulated") && !level.getBoolean("LightPopulated")) {
+                        net.minecraft.server.level.ServerChunkCache cps = (generatoraccess == null) ? null : generatoraccess.getChunkSource();
+                        if (check(cps, pos.x - 1, pos.z) && check(cps, pos.x - 1, pos.z - 1) && check(cps, pos.x, pos.z - 1)) {
+                            level.putBoolean("LightPopulated", true);
+                        }
+                    }
+                }
+                // CraftBukkit end
+
                 if (i < 1493) {
                     p_188291_ = DataFixTypes.CHUNK.update(this.fixerUpper, p_188291_, i, 1493);
                     if (p_188291_.getCompound("Level").getBoolean("hasLegacyStructureData")) {
