--- a/net/minecraft/world/level/chunk/storage/RegionFile.java
+++ b/net/minecraft/world/level/chunk/storage/RegionFile.java
@@ -82,6 +_,14 @@
                     if (l != 0) {
                         int i1 = getSectorNumber(l);
                         int j1 = getNumSectors(l);
+                        // Spigot start
+                        if (j1 == 255) {
+                            // We're maxed out, so we need to read the proper length from the section
+                            ByteBuffer realLen = ByteBuffer.allocate(4);
+                            this.file.read(realLen, i1 * 4096);
+                            j1 = (realLen.getInt(0) + 4) / 4096 + 1;
+                        }
+                        // Spigot end
                         if (i1 < 2) {
                             LOGGER.warn("Region file {} has invalid sector at index: {}; sector {} overlaps with header", p_63633_, k, i1);
                             this.offsets.put(k, 0);
@@ -118,6 +_,13 @@
             int j = getSectorNumber(i);
             int k = getNumSectors(i);
             int l = k * 4096;
+            // Spigot start
+            if (k == 255) {
+                ByteBuffer realLen = ByteBuffer.allocate(4);
+                this.file.read(realLen, j * 4096);
+                k = (realLen.getInt(0) + 4) / 4096 + 1;
+            }
+            // Spigot end
             ByteBuffer bytebuffer = ByteBuffer.allocate(l);
             this.file.read(bytebuffer, (long)(j * 4096));
             bytebuffer.flip();
@@ -376,6 +_,7 @@
         }
     }
 
+    public static final int MAX_CHUNK_SIZE = 500 * 1024 * 1024; // Paper - don't write garbage data to disk if writing serialization fails
     class ChunkBuffer extends ByteArrayOutputStream {
         private final ChunkPos pos;
 
@@ -388,6 +_,24 @@
             super.write(RegionFile.this.version.getId());
             this.pos = p_63696_;
         }
+
+        // Paper start - don't write garbage data to disk if writing serialization fails
+        @Override
+        public void write(final int b) {
+            if (this.count > MAX_CHUNK_SIZE) {
+                throw new RegionFileStorage.RegionFileSizeException("Region file too large: " + this.count);
+            }
+            super.write(b);
+        }
+
+        @Override
+        public void write(final byte[] b, final int off, final int len) {
+            if (this.count + len > MAX_CHUNK_SIZE) {
+                throw new RegionFileStorage.RegionFileSizeException("Region file too large: " + (this.count + len));
+            }
+            super.write(b, off, len);
+        }
+        // Paper end - don't write garbage data to disk if writing serialization fails
 
         @Override
         public void close() throws IOException {
