--- a/net/minecraft/world/entity/monster/Vindicator.java
+++ b/net/minecraft/world/entity/monster/Vindicator.java
@@ -179,7 +_,7 @@
 
     static class VindicatorBreakDoorGoal extends BreakDoorGoal {
         public VindicatorBreakDoorGoal(Mob p_34112_) {
-            super(p_34112_, 6, Vindicator.DOOR_BREAKING_PREDICATE);
+            super(p_34112_, 6, com.google.common.base.Predicates.in(p_34112_.level().paperConfig().entities.behavior.doorBreakingDifficulty.getOrDefault(p_34112_.getType(), p_34112_.level().paperConfig().entities.behavior.doorBreakingDifficulty.get(EntityType.VINDICATOR)))); // Paper - Configurable door breaking difficulty
             this.setFlags(EnumSet.of(Goal.Flag.MOVE));
         }
 
