--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -199,12 +_,62 @@
     private final List<CustomSpawner> customSpawners;
     @Nullable
     private EndDragonFight dragonFight;
-    final Int2ObjectMap<EnderDragonPart> dragonParts = new Int2ObjectOpenHashMap<>();
+    final Int2ObjectMap<net.neoforged.neoforge.entity.PartEntity<?>> dragonParts = new Int2ObjectOpenHashMap<>();
     private final StructureManager structureManager;
     private final StructureCheck structureCheck;
     private final boolean tickTime;
     private final RandomSequences randomSequences;
 
+    // CraftBukkit start
+    public net.minecraft.world.level.storage.PrimaryLevelData worldDataServer;
+    public final LevelStorageSource.LevelStorageAccess convertable;
+    public final UUID uuid;
+    public boolean hasPhysicsEvent = true; // Paper - BlockPhysicsEvent
+    public ResourceKey<LevelStem> typeKey;
+
+    public LevelChunk getChunkIfLoaded(int x, int z) {
+        return this.chunkSource.getChunk(x, z, false);
+    }
+
+    @Override
+    public ResourceKey<LevelStem> getTypeKey() {
+        return typeKey == null ? super.getTypeKey() : typeKey;
+    }
+
+    // Paper start
+    public final boolean areChunksLoadedForMove(AABB axisalignedbb) {
+        // copied code from collision methods, so that we can guarantee that they wont load chunks (we don't override
+        // ICollisionAccess methods for VoxelShapes)
+        // be more strict too, add a block (dumb plugins in move events?)
+        int minBlockX = Mth.floor(axisalignedbb.minX - 1.0E-7D) - 3;
+        int maxBlockX = Mth.floor(axisalignedbb.maxX + 1.0E-7D) + 3;
+        int minBlockZ = Mth.floor(axisalignedbb.minZ - 1.0E-7D) - 3;
+        int maxBlockZ = Mth.floor(axisalignedbb.maxZ + 1.0E-7D) + 3;
+        int minChunkX = minBlockX >> 4;
+        int maxChunkX = maxBlockX >> 4;
+        int minChunkZ = minBlockZ >> 4;
+        int maxChunkZ = maxBlockZ >> 4;
+        ServerChunkCache chunkProvider = this.getChunkSource();
+        for (int cx = minChunkX; cx <= maxChunkX; ++cx) {
+            for (int cz = minChunkZ; cz <= maxChunkZ; ++cz) {
+                if (chunkProvider.getChunkAtIfLoadedImmediately(cx, cz) == null) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+    // Paper end
+
+    // Paper start - optimise getPlayerByUUID
+    @Nullable
+    @Override
+    public Player getPlayerByUUID(UUID uuid) {
+        final Player player = this.getServer().getPlayerList().getPlayer(uuid);
+        return player != null && player.level() == this ? player : null;
+    }
+    // Paper end - optimise getPlayerByUUID
+
     public ServerLevel(
         MinecraftServer p_214999_,
         Executor p_215000_,
@@ -230,11 +_,66 @@
             p_215007_,
             p_214999_.getMaxChainedNeighborUpdates()
         );
+        // CraftBukkit start
+        this.pvpMode = p_214999_.isPvpAllowed();
+        convertable = p_215001_;
+        java.io.File worldFile = net.minecraft.world.level.dimension.DimensionType.getStorageFolder(p_215003_, p_215001_.levelDirectory.path()).toFile();
+        uuid = org.bukkit.craftbukkit.util.WorldUUID.getUUID(worldFile);
+        worldName =  worldFile.getName();
+        var typeKey = p_215001_.dimensionType;
+        if (typeKey != null) {
+            this.typeKey = typeKey;
+        } else {
+            var dimensions = p_214999_.registryAccess().registryOrThrow(Registries.LEVEL_STEM);
+            var key = dimensions.getResourceKey(p_215004_);
+            if (key.isPresent()) {
+                this.typeKey = key.get();
+            } else {
+                this.typeKey = ResourceKey.create(Registries.LEVEL_STEM, dimension.location());
+            }
+        }
+        // CraftBukkit end
         this.tickTime = p_215009_;
         this.server = p_214999_;
-        this.customSpawners = p_215008_;
         this.serverLevelData = p_215002_;
         ChunkGenerator chunkgenerator = p_215004_.generator();
+        // CraftBukkit start
+        //Magma start
+        if (p_215002_ instanceof net.minecraft.world.level.storage.PrimaryLevelData data) {
+            this.worldDataServer = data;
+        } else if (p_215002_ instanceof net.minecraft.world.level.storage.DerivedLevelData derived) {
+            this.worldDataServer = org.magmafoundation.magma.helper.DelegateWorldInfo.wrap(derived);
+        }
+
+        //Magma end
+        if (biomeProvider != null) {
+            org.bukkit.generator.WorldInfo worldInfo = new org.bukkit.craftbukkit.generator.CraftWorldInfo(worldDataServer, p_215001_, environment, p_215004_.type().value());
+            net.minecraft.world.level.biome.BiomeSource worldChunkManager = new org.bukkit.craftbukkit.generator.CustomWorldChunkManager(worldInfo, biomeProvider, server.registryAccess().registryOrThrow(Registries.BIOME));
+            if (chunkgenerator instanceof net.minecraft.world.level.levelgen.NoiseBasedChunkGenerator cga) {
+                chunkgenerator = new net.minecraft.world.level.levelgen.NoiseBasedChunkGenerator(worldChunkManager, cga.settings);
+            } else if (chunkgenerator instanceof net.minecraft.world.level.levelgen.FlatLevelSource cpf) {
+                chunkgenerator = new net.minecraft.world.level.levelgen.FlatLevelSource(cpf.settings(), worldChunkManager);
+            }
+        } else {
+            biomeProvider = getCraftServer().getBiomeProvider(p_215002_.getLevelName());
+        }
+        if (generator != null) {
+            chunkgenerator = new org.bukkit.craftbukkit.generator.CustomChunkGenerator(this, chunkgenerator, generator);
+        }else {
+            generator = getCraftServer().getGenerator(p_215002_.getLevelName());
+        }
+
+        // CraftBukkit end
+        this.spigotConfig = new org.spigotmc.SpigotWorldConfig(worldName); // Spigot
+        paperConfigFunction = spigotConfig -> p_214999_.paperConfigurations.createWorldConfig(
+                io.papermc.paper.configuration.PaperConfigurations.createWorldContextMap(
+                        p_215001_.levelDirectory.path(),
+                        worldName,
+                        p_215003_.location(),
+                        spigotConfig,
+                        p_214999_.registryAccess(),
+                        (worldDataServer != null ? worldDataServer.getGameRules() : serverLevelData.getGameRules())));
+        this.paperConfig = paperConfigFunction.apply(this.spigotConfig); // Paper - create paper world configs
         boolean flag = p_214999_.forceSynchronousWrites();
         DataFixer datafixer = p_214999_.getFixerUpper();
         EntityPersistentStorage<Entity> entitypersistentstorage = new EntityStorage(
@@ -256,8 +_,8 @@
             p_214999_.getStructureManager(),
             p_215000_,
             chunkgenerator,
-            p_214999_.getPlayerList().getViewDistance(),
-            p_214999_.getPlayerList().getSimulationDistance(),
+            this.spigotConfig.viewDistance, // Spigot
+            this.spigotConfig.simulationDistance, // Spigot
             flag,
             p_215005_,
             this.entityManager::updateChunkStatus,
@@ -275,16 +_,16 @@
 
         long i = p_214999_.getWorldData().worldGenOptions().seed();
         this.structureCheck = new StructureCheck(
-            this.chunkSource.chunkScanner(),
-            this.registryAccess(),
-            p_214999_.getStructureManager(),
-            p_215003_,
-            chunkgenerator,
-            this.chunkSource.randomState(),
-            this,
-            chunkgenerator.getBiomeSource(),
-            i,
-            datafixer
+                this.chunkSource.chunkScanner(),
+                this.registryAccess(),
+                p_214999_.getStructureManager(),
+                p_215003_,
+                chunkgenerator,
+                this.chunkSource.randomState(),
+                this,
+                chunkgenerator.getBiomeSource(),
+                i,
+                datafixer
         );
         this.structureManager = new StructureManager(this, p_214999_.getWorldData().worldGenOptions(), this.structureCheck);
         if (this.dimension() == Level.END && this.dimensionTypeRegistration().is(BuiltinDimensionTypes.END)) {
@@ -298,7 +_,27 @@
         this.randomSequences = Objects.requireNonNullElseGet(
             p_288977_, () -> this.getDataStorage().computeIfAbsent(RandomSequences.factory(i), "random_sequences")
         );
-    }
+
+        net.neoforged.neoforge.attachment.LevelAttachmentsSavedData.init(this);
+        // Neo: Move the initialization of customSpawners to the end of costructor
+        // Providing a fully initialized ServerLevel instance for the ServerLevelEvent.CustomSpawners
+        this.customSpawners = net.neoforged.neoforge.event.EventHooks.getCustomSpawners(this, p_215008_);
+        this.world = new org.bukkit.craftbukkit.CraftWorld(this, generator, biomeProvider, environment);
+        this.getCraftServer().addWorld(this.getWorld()); // CraftBukkit
+        if (worldDataServer != null) {
+            worldDataServer.setWorld(this);
+        }
+        this.timings = new co.aikar.timings.WorldTimingsHandler(this); // Paper - code below can generate new world and access timings
+        this.entityLimiter = new org.spigotmc.TickLimiter(spigotConfig.entityMaxTickTime);
+        this.tileLimiter = new org.spigotmc.TickLimiter(spigotConfig.tileMaxTickTime);
+    }
+
+    // Paper start
+    @Override
+    public boolean hasChunk(int chunkX, int chunkZ) {
+        return this.getChunkSource().getChunkAtIfLoadedImmediately(chunkX, chunkZ) != null;
+    }
+    // Paper end
 
     @Deprecated
     @VisibleForTesting
@@ -342,7 +_,7 @@
         if (this.sleepStatus.areEnoughSleeping(i) && this.sleepStatus.areEnoughDeepSleeping(i, this.players)) {
             if (this.getGameRules().getBoolean(GameRules.RULE_DAYLIGHT)) {
                 long j = this.levelData.getDayTime() + 24000L;
-                this.setDayTime(j - j % 24000L);
+                this.setDayTime(net.neoforged.neoforge.event.EventHooks.onSleepFinished(this, j - j % 24000L, this.getDayTime()));
             }
 
             this.wakeUpAllPlayers();
@@ -357,6 +_,7 @@
         }
 
         profilerfiller.popPush("tickPending");
+        this.timings.scheduledBlocks.startTiming(); // Paper
         if (!this.isDebug() && flag) {
             long k = this.getGameTime();
             profilerfiller.push("blockTicks");
@@ -365,34 +_,44 @@
             this.fluidTicks.tick(k, 65536, this::tickFluid);
             profilerfiller.pop();
         }
+        this.timings.scheduledBlocks.stopTiming(); // Paper
 
         profilerfiller.popPush("raid");
         if (flag) {
+            this.timings.raids.startTiming(); // Paper - timings
             this.raids.tick();
+            this.timings.raids.stopTiming(); // Paper - timings
         }
 
         profilerfiller.popPush("chunkSource");
+        this.timings.chunkProviderTick.startTiming(); // Paper - timings
         this.getChunkSource().tick(p_8794_, true);
+        this.timings.chunkProviderTick.stopTiming(); // Paper - timings
         profilerfiller.popPush("blockEvents");
         if (flag) {
+            timings.doSounds.startTiming(); // Spigot
             this.runBlockEvents();
+            timings.doSounds.stopTiming(); // Spigot
         }
 
         this.handlingTick = false;
         profilerfiller.pop();
-        boolean flag1 = !this.players.isEmpty() || !this.getForcedChunks().isEmpty();
+        boolean flag1 = !this.players.isEmpty() || net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.hasForcedChunks(this); // Neo: Replace vanilla's has forced chunk check with neo's that checks both the vanilla and neo added ones
         if (flag1) {
             this.resetEmptyTime();
         }
 
         if (flag1 || this.emptyTime++ < 300) {
             profilerfiller.push("entities");
+            timings.tickEntities.startTiming(); // Spigot
             if (this.dragonFight != null && flag) {
                 profilerfiller.push("dragonFight");
                 this.dragonFight.tick();
                 profilerfiller.pop();
             }
 
+            org.spigotmc.ActivationRange.activateEntities(this); // Spigot
+            timings.entityTick.startTiming(); // Spigot
             this.entityTickList.forEach(p_308566_ -> {
                 if (!p_308566_.isRemoved()) {
                     if (this.shouldDiscardEntity(p_308566_)) {
@@ -412,12 +_,16 @@
                             }
 
                             profilerfiller.push("tick");
-                            this.guardEntityTick(this::tickNonPassenger, p_308566_);
+                            if (!p_308566_.isRemoved() && !(p_308566_ instanceof net.neoforged.neoforge.entity.PartEntity)) {
+                                this.guardEntityTick(this::tickNonPassenger, p_308566_);
+                            }
                             profilerfiller.pop();
                         }
                     }
                 }
             });
+            timings.entityTick.stopTiming(); // Spigot
+            timings.tickEntities.stopTiming(); // Spigot
             profilerfiller.pop();
             this.tickBlockEntities();
         }
@@ -438,7 +_,7 @@
             this.serverLevelData.setGameTime(i);
             this.serverLevelData.getScheduledEvents().tick(this.server, i);
             if (this.levelData.getGameRules().getBoolean(GameRules.RULE_DAYLIGHT)) {
-                this.setDayTime(this.levelData.getDayTime() + 1L);
+                this.setDayTime(this.levelData.getDayTime() + advanceDaytime());
             }
         }
     }
@@ -471,20 +_,20 @@
         int j = chunkpos.getMinBlockZ();
         ProfilerFiller profilerfiller = this.getProfiler();
         profilerfiller.push("thunder");
-        if (flag && this.isThundering() && this.random.nextInt(100000) == 0) {
+        if (!this.paperConfig().environment.disableThunder && flag && this.isThundering() && this.spigotConfig.thunderChance > 0 && this.random.nextInt(this.spigotConfig.thunderChance) == 0) { // Spigot // Paper - Option to disable thunder
             BlockPos blockpos = this.findLightningTargetAround(this.getBlockRandomPos(i, 0, j, 15));
             if (this.isRainingAt(blockpos)) {
                 DifficultyInstance difficultyinstance = this.getCurrentDifficultyAt(blockpos);
                 boolean flag1 = this.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)
-                    && this.random.nextDouble() < (double)difficultyinstance.getEffectiveDifficulty() * 0.01
-                    && !this.getBlockState(blockpos.below()).is(Blocks.LIGHTNING_ROD);
+                    && this.random.nextDouble() < (double)difficultyinstance.getEffectiveDifficulty() * this.paperConfig().entities.spawning.skeletonHorseThunderSpawnChance.or(0.01D) // Paper - Configurable spawn chances for skeleton horses
+                        && !(this.getBlockState(blockpos.below()).getBlock() instanceof net.minecraft.world.level.block.LightningRodBlock); // Neo: support custom LightningRodBlocks
                 if (flag1) {
                     SkeletonHorse skeletonhorse = EntityType.SKELETON_HORSE.create(this);
                     if (skeletonhorse != null) {
                         skeletonhorse.setTrap(true);
                         skeletonhorse.setAge(0);
                         skeletonhorse.setPos((double)blockpos.getX(), (double)blockpos.getY(), (double)blockpos.getZ());
-                        this.addFreshEntity(skeletonhorse);
+                        this.addFreshEntity(skeletonhorse, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING); // CraftBukkit
                     }
                 }
 
@@ -492,20 +_,23 @@
                 if (lightningbolt != null) {
                     lightningbolt.moveTo(Vec3.atBottomCenterOf(blockpos));
                     lightningbolt.setVisualOnly(flag1);
-                    this.addFreshEntity(lightningbolt);
+                    this.strikeLightning(lightningbolt, org.bukkit.event.weather.LightningStrikeEvent.Cause.WEATHER); // CraftBukkit
                 }
             }
         }
 
         profilerfiller.popPush("iceandsnow");
 
-        for (int i1 = 0; i1 < p_8716_; i1++) {
-            if (this.random.nextInt(48) == 0) {
-                this.tickPrecipitation(this.getBlockRandomPos(i, 0, j, 15));
+        if (!this.paperConfig().environment.disableIceAndSnow) { // Paper - Option to disable ice and snow
+            for (int i1 = 0; i1 < p_8716_; i1++) {
+                if (this.random.nextInt(48) == 0) {
+                    this.tickPrecipitation(this.getBlockRandomPos(i, 0, j, 15));
+                }
             }
-        }
+        } // Paper - Option to disable ice and snow
 
         profilerfiller.popPush("tickBlocks");
+        timings.chunkTicksBlocks.startTiming(); // Paper
         if (p_8716_ > 0) {
             LevelChunkSection[] alevelchunksection = p_8715_.getSections();
 
@@ -534,6 +_,7 @@
             }
         }
 
+        timings.chunkTicksBlocks.stopTiming(); // Paper
         profilerfiller.pop();
     }
 
@@ -542,8 +_,9 @@
         BlockPos blockpos = this.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, p_295060_);
         BlockPos blockpos1 = blockpos.below();
         Biome biome = this.getBiome(blockpos).value();
+        if (this.isAreaLoaded(blockpos1, 1)) // Forge: check area to avoid loading neighbors in unloaded chunks
         if (biome.shouldFreeze(this, blockpos1)) {
-            this.setBlockAndUpdate(blockpos1, Blocks.ICE.defaultBlockState());
+            org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(this, blockpos1, Blocks.ICE.defaultBlockState(), null); // CraftBukkit
         }
 
         if (this.isRaining()) {
@@ -555,10 +_,10 @@
                     if (j < Math.min(i, 8)) {
                         BlockState blockstate1 = blockstate.setValue(SnowLayerBlock.LAYERS, Integer.valueOf(j + 1));
                         Block.pushEntitiesUp(blockstate, blockstate1, this, blockpos);
-                        this.setBlockAndUpdate(blockpos, blockstate1);
+                        org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(this, blockpos, blockstate1, null); // CraftBukkit
                     }
                 } else {
-                    this.setBlockAndUpdate(blockpos, Blocks.SNOW.defaultBlockState());
+                    org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(this, blockpos, Blocks.SNOW.defaultBlockState(), null); // CraftBukkit
                 }
             }
 
@@ -714,15 +_,19 @@
                 .broadcastAll(new ClientboundGameEventPacket(ClientboundGameEventPacket.THUNDER_LEVEL_CHANGE, this.thunderLevel), this.dimension());
         }
 
+        /* The function in use here has been replaced in order to only send the weather info to players in the correct dimension,
+         * rather than to all players on the server. This is what causes the client-side rain, as the
+         * client believes that it has started raining locally, rather than in another dimension.
+         */
         if (flag != this.isRaining()) {
             if (flag) {
-                this.server.getPlayerList().broadcastAll(new ClientboundGameEventPacket(ClientboundGameEventPacket.STOP_RAINING, 0.0F));
+                this.server.getPlayerList().broadcastAll(new ClientboundGameEventPacket(ClientboundGameEventPacket.STOP_RAINING, 0.0F), this.dimension());
             } else {
-                this.server.getPlayerList().broadcastAll(new ClientboundGameEventPacket(ClientboundGameEventPacket.START_RAINING, 0.0F));
+                this.server.getPlayerList().broadcastAll(new ClientboundGameEventPacket(ClientboundGameEventPacket.START_RAINING, 0.0F), this.dimension());
             }
 
-            this.server.getPlayerList().broadcastAll(new ClientboundGameEventPacket(ClientboundGameEventPacket.RAIN_LEVEL_CHANGE, this.rainLevel));
-            this.server.getPlayerList().broadcastAll(new ClientboundGameEventPacket(ClientboundGameEventPacket.THUNDER_LEVEL_CHANGE, this.thunderLevel));
+            this.server.getPlayerList().broadcastAll(new ClientboundGameEventPacket(ClientboundGameEventPacket.RAIN_LEVEL_CHANGE, this.rainLevel), this.dimension());
+            this.server.getPlayerList().broadcastAll(new ClientboundGameEventPacket(ClientboundGameEventPacket.THUNDER_LEVEL_CHANGE, this.thunderLevel), this.dimension());
         }
     }
 
@@ -753,17 +_,39 @@
     }
 
     public void tickNonPassenger(Entity p_8648_) {
-        p_8648_.setOldPosAndRot();
-        ProfilerFiller profilerfiller = this.getProfiler();
-        p_8648_.tickCount++;
-        this.getProfiler().push(() -> BuiltInRegistries.ENTITY_TYPE.getKey(p_8648_.getType()).toString());
-        profilerfiller.incrementCounter("tickNonPassenger");
-        p_8648_.tick();
-        this.getProfiler().pop();
-
-        for (Entity entity : p_8648_.getPassengers()) {
-            this.tickPassenger(p_8648_, entity);
+        ++co.aikar.timings.TimingHistory.entityTicks; // Paper - timings
+        // Spigot start
+        co.aikar.timings.Timing timer; // Paper
+        if (!org.spigotmc.ActivationRange.checkIfActive(p_8648_)) {
+            p_8648_.tickCount++;
+            timer = entity.getType().inactiveTickTimer.startTiming(); try { // Paper - timings
+            p_8648_.inactiveTick();
+            } finally { timer.stopTiming(); } // Paper
+            return;
         }
+        // Spigot end
+        // Paper start- timings
+        co.aikar.timings.TimingHistory.activatedEntityTicks++;
+        timer = p_8648_.getVehicle() != null ? p_8648_.getType().passengerTickTimer.startTiming() : p_8648_.getType().tickTimer.startTiming();
+        try {
+            // Paper end - timings
+            p_8648_.setOldPosAndRot();
+            ProfilerFiller profilerfiller = this.getProfiler();
+            p_8648_.tickCount++;
+            this.getProfiler().push(() -> BuiltInRegistries.ENTITY_TYPE.getKey(p_8648_.getType()).toString());
+            profilerfiller.incrementCounter("tickNonPassenger");
+            // Neo: Permit cancellation of Entity#tick via EntityTickEvent.Pre
+            if (!net.neoforged.neoforge.event.EventHooks.fireEntityTickPre(p_8648_).isCanceled()) {
+                p_8648_.tick();
+                p_8648_.postTick(); // CraftBukkit
+                net.neoforged.neoforge.event.EventHooks.fireEntityTickPost(p_8648_);
+            }
+            this.getProfiler().pop();
+
+            for (Entity entity : p_8648_.getPassengers()) {
+                this.tickPassenger(p_8648_, entity);
+            }
+        } finally { timer.stopTiming(); } // Paper - timings
     }
 
     private void tickPassenger(Entity p_8663_, Entity p_8664_) {
@@ -793,13 +_,18 @@
         ServerChunkCache serverchunkcache = this.getChunkSource();
         if (!p_8646_) {
             if (p_8644_ != null) {
-                p_8644_.progressStartNoAbort(Component.translatable("menu.savingLevel"));
-            }
+                org.bukkit.Bukkit.getPluginManager().callEvent(new org.bukkit.event.world.WorldSaveEvent(getWorld())); // CraftBukkit
+                try (co.aikar.timings.Timing ignored = timings.worldSave.startTiming()) { // Paper
+                    p_8644_.progressStartNoAbort(Component.translatable("menu.savingLevel"));
+                }
 
-            this.saveLevelData();
-            if (p_8644_ != null) {
-                p_8644_.progressStage(Component.translatable("menu.savingChunks"));
-            }
+                timings.worldSaveChunks.startTiming(); // Paper
+                this.saveLevelData();
+                if (p_8644_ != null) {
+                    p_8644_.progressStage(Component.translatable("menu.savingChunks"));
+                }
+                timings.worldSaveChunks.stopTiming(); // Paper
+            }// Paper
 
             serverchunkcache.save(p_8645_);
             if (p_8645_) {
@@ -807,6 +_,18 @@
             } else {
                 this.entityManager.autoSave();
             }
+
+            // CraftBukkit start - moved from MinecraftServer.saveChunks
+            ServerLevel worldserver1 = this;
+            worldDataServer.setWorldBorder(worldserver1.getWorldBorder().createSettings());
+            worldDataServer.setCustomBossEvents(this.server.getCustomBossEvents().save(this.registryAccess()));
+            convertable.saveDataTag(this.server.registryAccess(), this.worldDataServer, this.server.getPlayerList().getSingleplayerData());
+            // CraftBukkit end
+
+            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.LevelEvent.Save(this));
+            if (p_8645_) {
+                net.neoforged.neoforge.common.IOUtilities.waitUntilIOWorkerComplete();
+            }
         }
     }
 
@@ -870,19 +_,48 @@
         return list.isEmpty() ? null : list.get(this.random.nextInt(list.size()));
     }
 
+    // Magma
+    private org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason magma$spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT;
+    public void setSpawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+        magma$spawnReason = reason;
+    }
+
     @Override
     public boolean addFreshEntity(Entity p_8837_) {
-        return this.addEntity(p_8837_);
+        // CraftBukkit start
+        return this.addFreshEntity(p_8837_, magma$spawnReason);
+    }
+
+    @Override
+    public boolean addFreshEntity(Entity entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+        return this.addEntity(entity);
+        // CraftBukkit end
     }
 
     public boolean addWithUUID(Entity p_8848_) {
+        // CraftBukkit start
+        return this.addWithUUID(p_8848_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+    }
+
+    public boolean addWithUUID(Entity p_8848_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
         return this.addEntity(p_8848_);
+        // CraftBukkit end
     }
 
     public void addDuringTeleport(Entity p_143335_) {
+        // CraftBukkit start
+        // SPIGOT-6415: Don't call spawn event for entities which travel trough worlds,
+        // since it is only an implementation detail, that a new entity is created when
+        // they are traveling between worlds.
+        this.addDuringTeleport(p_143335_, null);
+    }
+
+    public void addDuringTeleport(Entity p_143335_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+        // CraftBukkit end
         if (p_143335_ instanceof ServerPlayer serverplayer) {
             this.addPlayer(serverplayer);
         } else {
+            this.setSpawnReason(reason);
             this.addEntity(p_143335_);
         }
     }
@@ -896,6 +_,7 @@
     }
 
     private void addPlayer(ServerPlayer p_8854_) {
+        if (net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.EntityJoinLevelEvent(p_8854_, this)).isCanceled()) return;
         Entity entity = this.getEntities().get(p_8854_.getUUID());
         if (entity != null) {
             LOGGER.warn("Force-added player with duplicate UUID {}", p_8854_.getUUID());
@@ -903,28 +_,63 @@
             this.removePlayerImmediately((ServerPlayer)entity, Entity.RemovalReason.DISCARDED);
         }
 
-        this.entityManager.addNewEntity(p_8854_);
+        this.entityManager.addNewEntityWithoutEvent(p_8854_);
+        p_8854_.onAddedToLevel();
     }
 
     private boolean addEntity(Entity p_8873_) {
+        org.spigotmc.AsyncCatcher.catchOp("entity add"); // Spigot
+        // Paper start - extra debug info
+        if (p_8873_.valid) {
+            LOGGER.error("Attempted Double World add on {}", p_8873_, new Throwable());
+            return true;
+        }
+        // Paper end - extra debug info
+        if (p_8873_.spawnReason == null) p_8873_.spawnReason = magma$spawnReason; // Paper - Entity#getEntitySpawnReason
         if (p_8873_.isRemoved()) {
-            LOGGER.warn("Tried to add entity {} but it was marked as removed already", EntityType.getKey(p_8873_.getType()));
+            // LOGGER.warn("Tried to add entity {} but it was marked as removed already", EntityType.getKey(p_8873_.getType())); // CraftBukkit
             return false;
         } else {
-            return this.entityManager.addNewEntity(p_8873_);
+            // SPIGOT-6415: Don't call spawn event when reason is null. For example when an entity teleports to a new world.
+            if (magma$spawnReason != null && !org.bukkit.craftbukkit.event.CraftEventFactory.doEntityAddEventCalling(this, p_8873_, magma$spawnReason)) {
+                return false;
+            }
+            // CraftBukkit end
+
+            if (this.entityManager.addNewEntity(p_8873_)) {
+                p_8873_.onAddedToLevel();
+                return true;
+            } else {
+                return false;
+            }
         }
     }
 
     public boolean tryAddFreshEntityWithPassengers(Entity p_8861_) {
+        // CraftBukkit start
+        return this.tryAddFreshEntityWithPassengers(p_8861_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+    }
+
+    public boolean tryAddFreshEntityWithPassengers(Entity p_8861_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+        // CraftBukkit end
         if (p_8861_.getSelfAndPassengers().map(Entity::getUUID).anyMatch(this.entityManager::isLoaded)) {
             return false;
         } else {
-            this.addFreshEntityWithPassengers(p_8861_);
+            this.addFreshEntityWithPassengers(p_8861_, reason); // CraftBukkit
             return true;
         }
     }
 
     public void unload(LevelChunk p_8713_) {
+        // Spigot Start
+        for (net.minecraft.world.level.block.entity.BlockEntity tileentity : p_8713_.getBlockEntities().values()) {
+            if (tileentity instanceof net.minecraft.world.Container) {
+                for (org.bukkit.entity.HumanEntity h : Lists.newArrayList(((net.minecraft.world.Container) tileentity).getViewers())) {
+                    h.closeInventory(org.bukkit.event.inventory.InventoryCloseEvent.Reason.UNLOADED); // Paper - Inventory close reason
+                }
+            }
+        }
+        // Spigot End
         p_8713_.clearAllBlockEntities();
         p_8713_.unregisterTickContainerFromLevel(this);
     }
@@ -933,13 +_,41 @@
         p_143262_.remove(p_143263_);
     }
 
+    // CraftBukkit start
+    public boolean strikeLightning(Entity entitylightning) {
+        return this.strikeLightning(entitylightning, org.bukkit.event.weather.LightningStrikeEvent.Cause.UNKNOWN);
+    }
+
+    public boolean strikeLightning(Entity entitylightning, org.bukkit.event.weather.LightningStrikeEvent.Cause cause) {
+        org.bukkit.event.weather.LightningStrikeEvent lightning = org.bukkit.craftbukkit.event.CraftEventFactory.callLightningStrikeEvent((org.bukkit.entity.LightningStrike) entitylightning.getBukkitEntity(), cause);
+        if (lightning.isCancelled()) {
+            return false;
+        }
+        return this.addFreshEntity(entitylightning);
+    }
+    // CraftBukkit end
+
     @Override
     public void destroyBlockProgress(int p_8612_, BlockPos p_8613_, int p_8614_) {
+
+        // CraftBukkit start
+        Player entityhuman = null;
+        Entity entity = this.getEntity(p_8612_);
+        if (entity instanceof Player) entityhuman = (Player) entity;
+        // CraftBukkit end
+
         for (ServerPlayer serverplayer : this.server.getPlayerList().getPlayers()) {
             if (serverplayer != null && serverplayer.level() == this && serverplayer.getId() != p_8612_) {
                 double d0 = (double)p_8613_.getX() - serverplayer.getX();
                 double d1 = (double)p_8613_.getY() - serverplayer.getY();
                 double d2 = (double)p_8613_.getZ() - serverplayer.getZ();
+
+                // CraftBukkit start
+                if (entityhuman != null && !serverplayer.getBukkitEntity().canSee(entityhuman.getBukkitEntity())) {
+                    continue;
+                }
+                // CraftBukkit end
+
                 if (d0 * d0 + d1 * d1 + d2 * d2 < 1024.0) {
                     serverplayer.connection.send(new ClientboundBlockDestructionPacket(p_8612_, p_8613_, p_8614_));
                 }
@@ -959,6 +_,12 @@
         float p_263390_,
         long p_263403_
     ) {
+        net.neoforged.neoforge.event.PlayLevelSoundEvent.AtPosition event = net.neoforged.neoforge.event.EventHooks.onPlaySoundAtPosition(this, p_263393_, p_263369_, p_263354_, p_263412_, p_263338_, p_263352_, p_263390_);
+        if (event.isCanceled() || event.getSound() == null) return;
+        p_263412_ = event.getSound();
+        p_263338_ = event.getSource();
+        p_263352_ = event.getNewVolume();
+        p_263390_ = event.getNewPitch();
         this.server
             .getPlayerList()
             .broadcast(
@@ -976,6 +_,12 @@
     public void playSeededSound(
         @Nullable Player p_263545_, Entity p_263544_, Holder<SoundEvent> p_263491_, SoundSource p_263542_, float p_263530_, float p_263520_, long p_263490_
     ) {
+        net.neoforged.neoforge.event.PlayLevelSoundEvent.AtEntity event = net.neoforged.neoforge.event.EventHooks.onPlaySoundAtEntity(p_263544_, p_263491_, p_263542_, p_263530_, p_263520_);
+        if (event.isCanceled() || event.getSound() == null) return;
+        p_263491_ = event.getSound();
+        p_263542_ = event.getSource();
+        p_263530_ = event.getNewVolume();
+        p_263520_ = event.getNewPitch();
         this.server
             .getPlayerList()
             .broadcast(
@@ -1007,7 +_,7 @@
                 (double)p_8686_.getX(),
                 (double)p_8686_.getY(),
                 (double)p_8686_.getZ(),
-                64.0,
+                64.0, // Paper - diff on change (the 64.0 distance is used as defaults for sound ranges in spigot config for ender dragon, end portal and wither)
                 this.dimension(),
                 new ClientboundLevelEventPacket(p_8685_, p_8686_, p_8687_, false)
             );
@@ -1019,6 +_,7 @@
 
     @Override
     public void gameEvent(Holder<GameEvent> p_316597_, Vec3 p_215042_, GameEvent.Context p_215043_) {
+        if (!net.neoforged.neoforge.common.CommonHooks.onVanillaGameEvent(this, p_316597_, p_215042_, p_215043_)) return;
         this.gameEventDispatcher.post(p_316597_, p_215042_, p_215043_);
     }
 
@@ -1057,11 +_,17 @@
 
     @Override
     public void updateNeighborsAt(BlockPos p_215045_, Block p_215046_) {
+        if (captureBlockStates) { return; } // Paper - Cancel all physics during placement
+        net.neoforged.neoforge.event.EventHooks.onNeighborNotify(this, p_215045_, this.getBlockState(p_215045_), java.util.EnumSet.allOf(Direction.class), false).isCanceled();
         this.neighborUpdater.updateNeighborsAtExceptFromFacing(p_215045_, p_215046_, null);
     }
 
     @Override
     public void updateNeighborsAtExceptFromFacing(BlockPos p_215052_, Block p_215053_, Direction p_215054_) {
+        java.util.EnumSet<Direction> directions = java.util.EnumSet.allOf(Direction.class);
+        directions.remove(p_215054_);
+        if (net.neoforged.neoforge.event.EventHooks.onNeighborNotify(this, p_215052_, this.getBlockState(p_215052_), directions, false).isCanceled())
+            return;
         this.neighborUpdater.updateNeighborsAtExceptFromFacing(p_215052_, p_215053_, p_215054_);
     }
 
@@ -1089,6 +_,7 @@
         return this.chunkSource;
     }
 
+    // Paper start - Allow explosions to damage source
     @Override
     public Explosion explode(
         @Nullable Entity p_256039_,
@@ -1102,11 +_,18 @@
         Level.ExplosionInteraction p_255827_,
         ParticleOptions p_312436_,
         ParticleOptions p_312391_,
-        Holder<SoundEvent> p_320497_
+        Holder<SoundEvent> p_320497_,
+        java.util.function.Consumer<Explosion> configurator
     ) {
         Explosion explosion = this.explode(
-            p_256039_, p_255778_, p_256002_, p_256067_, p_256370_, p_256153_, p_256045_, p_255686_, p_255827_, false, p_312436_, p_312391_, p_320497_
+            p_256039_, p_255778_, p_256002_, p_256067_, p_256370_, p_256153_, p_256045_, p_255686_, p_255827_, false, p_312436_, p_312391_, p_320497_, configurator
         );
+        // Paper end - Allow explosions to damage source
+        // CraftBukkit start
+        if (explosion.wasCanceled) {
+            return explosion;
+        }
+        // CraftBukkit end
         if (!explosion.interactsWithBlocks()) {
             explosion.clearToBlow();
         }
@@ -1196,14 +_,25 @@
     public <T extends ParticleOptions> int sendParticles(
         T p_8768_, double p_8769_, double p_8770_, double p_8771_, int p_8772_, double p_8773_, double p_8774_, double p_8775_, double p_8776_
     ) {
-        ClientboundLevelParticlesPacket clientboundlevelparticlespacket = new ClientboundLevelParticlesPacket(
-            p_8768_, false, p_8769_, p_8770_, p_8771_, (float)p_8773_, (float)p_8774_, (float)p_8775_, (float)p_8776_, p_8772_
-        );
+        // CraftBukkit - visibility api support
+        return sendParticles(null, p_8768_, p_8769_, p_8770_, p_8771_, p_8772_, p_8773_, p_8774_, p_8775_, p_8776_, false);
+    }
+
+    public <T extends ParticleOptions> int sendParticles(ServerPlayer sender, T p_8768_, double p_8769_, double p_8770_, double p_8771_, int p_8772_, double p_8773_, double p_8774_, double p_8775_, double p_8776_, boolean force) {
+        // Paper start - Particle API
+        return sendParticles(players, sender, p_8768_, p_8769_, p_8770_, p_8771_, p_8772_, p_8773_, p_8774_, p_8775_, p_8776_, force);
+    }
+
+    public <T extends ParticleOptions> int sendParticles(List<ServerPlayer> receivers, @Nullable ServerPlayer sender, T p_8768_, double p_8769_, double p_8770_, double p_8771_, int p_8772_, double p_8773_, double p_8774_, double p_8775_, double p_8776_, boolean force) {
+        // Paper end - Particle API
+        ClientboundLevelParticlesPacket clientboundlevelparticlespacket = new ClientboundLevelParticlesPacket( p_8768_, force, p_8769_, p_8770_, p_8771_, (float)p_8773_, (float)p_8774_, (float)p_8775_, (float)p_8776_, p_8772_);
+        // CraftBukkit end
         int i = 0;
 
-        for (int j = 0; j < this.players.size(); j++) {
-            ServerPlayer serverplayer = this.players.get(j);
-            if (this.sendParticles(serverplayer, false, p_8769_, p_8770_, p_8771_, clientboundlevelparticlespacket)) {
+        for (Player entityhuman : receivers) { // Paper - Particle API
+            ServerPlayer serverplayer = (ServerPlayer) entityhuman; // Paper - Particle API
+            if (sender != null && !serverplayer.getBukkitEntity().canSee(sender.getBukkitEntity())) continue; // CraftBukkit
+            if (this.sendParticles(serverplayer, force, p_8769_, p_8770_, p_8771_, clientboundlevelparticlespacket)) { // CraftBukkit
                 i++;
             }
         }
@@ -1309,11 +_,22 @@
     @Nullable
     @Override
     public MapItemSavedData getMapData(MapId p_323746_) {
-        return this.getServer().overworld().getDataStorage().get(MapItemSavedData.factory(), p_323746_.key());
+        // CraftBukkit start
+        MapItemSavedData worldmap = (MapItemSavedData) this.getServer().overworld().getDataStorage().get(MapItemSavedData.factory(), p_323746_.key());
+        if (worldmap != null) {
+            worldmap.id = p_323746_;
+        }
+        return worldmap;
+        // CraftBukkit end
     }
 
     @Override
     public void setMapData(MapId p_323697_, MapItemSavedData p_143306_) {
+        // CraftBukkit start
+        p_143306_.id = p_323697_;
+        org.bukkit.event.server.MapInitializeEvent event = new org.bukkit.event.server.MapInitializeEvent(p_143306_.mapView);
+        org.bukkit.Bukkit.getServer().getPluginManager().callEvent(event);
+        // CraftBukkit end
         this.getServer().overworld().getDataStorage().set(p_323697_.key(), p_143306_);
     }
 
@@ -1525,6 +_,11 @@
     @Override
     public void blockUpdated(BlockPos p_8743_, Block p_8744_) {
         if (!this.isDebug()) {
+            // CraftBukkit start
+            if (populating) {
+                return;
+            }
+            // CraftBukkit end
             this.updateNeighborsAt(p_8743_, p_8744_);
         }
     }
@@ -1600,6 +_,7 @@
 
     @Override
     public LevelEntityGetter<Entity> getEntities() {
+        org.spigotmc.AsyncCatcher.catchOp("Chunk getEntities call"); // Spigot
         return this.entityManager.getEntityGetter();
     }
 
@@ -1693,10 +_,17 @@
 
         public void onTickingEnd(Entity p_143367_) {
             ServerLevel.this.entityTickList.remove(p_143367_);
+            // Paper start - Reset pearls when they stop being ticked
+            if (paperConfig().fixes.disableUnloadedChunkEnderpearlExploit && p_143367_ instanceof net.minecraft.world.entity.projectile.ThrownEnderpearl pearl) {
+                pearl.cachedOwner = null;
+                pearl.ownerUUID = null;
+            }
+            // Paper end - Reset pearls when they stop being ticked
         }
 
         public void onTrackingStart(Entity p_143371_) {
-            ServerLevel.this.getChunkSource().addEntity(p_143371_);
+            org.spigotmc.AsyncCatcher.catchOp("entity register"); // Spigot
+            // ServerLevel.this.getChunkSource().addEntity(p_143371_); // Paper - ignore and warn about illegal addEntity calls instead of crashing server; moved down below valid=true
             if (p_143371_ instanceof ServerPlayer serverplayer) {
                 ServerLevel.this.players.add(serverplayer);
                 ServerLevel.this.updateSleepingPlayerList();
@@ -1713,16 +_,60 @@
                 ServerLevel.this.navigatingMobs.add(mob);
             }
 
-            if (p_143371_ instanceof EnderDragon enderdragon) {
-                for (EnderDragonPart enderdragonpart : enderdragon.getSubEntities()) {
+            if (p_143371_.isMultipartEntity()) {
+                for(net.neoforged.neoforge.entity.PartEntity<?> enderdragonpart : p_143371_.getParts()) {
                     ServerLevel.this.dragonParts.put(enderdragonpart.getId(), enderdragonpart);
                 }
             }
 
             p_143371_.updateDynamicGameEventListener(DynamicGameEventListener::add);
+            p_143371_.inWorld = true; // CraftBukkit - Mark entity as in world
+            p_143371_.valid = true; // CraftBukkit
+            ServerLevel.this.getChunkSource().addEntity(p_143371_); // Paper - ignore and warn about illegal addEntity calls instead of crashing server
+            // Paper start - Entity origin API
+            if (p_143371_.getOriginVector() == null) {
+                p_143371_.setOrigin(p_143371_.getBukkitEntity().getLocation());
+            }
+            // Default to current world if unknown, gross assumption but entities rarely change world
+            if (p_143371_.getOriginWorld() == null) {
+                p_143371_.setOrigin(p_143371_.getOriginVector().toLocation(getWorld()));
+            }
+            // Paper end - Entity origin API
+            new com.destroystokyo.paper.event.entity.EntityAddToWorldEvent(p_143371_.getBukkitEntity(), ServerLevel.this.getWorld()).callEvent(); // Paper - fire while valid
         }
 
         public void onTrackingEnd(Entity p_143375_) {
+            org.spigotmc.AsyncCatcher.catchOp("entity unregister"); // Spigot
+            // Spigot start
+            if ( p_143375_ instanceof Player )
+            {
+                com.google.common.collect.Streams.stream( getServer().getAllLevels() ).map( ServerLevel::getDataStorage ).forEach( (worldData) ->
+                {
+                    for (Object o : worldData.cache.values() )
+                    {
+                        if ( o instanceof MapItemSavedData )
+                        {
+                            MapItemSavedData map = (MapItemSavedData) o;
+                            map.carriedByPlayers.remove( (Player) p_143375_ );
+                            for ( java.util.Iterator<net.minecraft.world.level.saveddata.maps.MapItemSavedData.HoldingPlayer> iter = (java.util.Iterator<net.minecraft.world.level.saveddata.maps.MapItemSavedData.HoldingPlayer>) map.carriedBy.iterator(); iter.hasNext(); )
+                            {
+                                if ( iter.next().player == p_143375_ )
+                                {
+                                    iter.remove();
+                                }
+                            }
+                        }
+                    }
+                } );
+            }
+            // Spigot end
+            // Spigot Start
+            if (p_143375_.getBukkitEntity() instanceof org.bukkit.inventory.InventoryHolder && (!(p_143375_ instanceof ServerPlayer) || p_143375_.getRemovalReason() != Entity.RemovalReason.KILLED)) { // SPIGOT-6876: closeInventory clears death message
+                for (org.bukkit.entity.HumanEntity h : Lists.newArrayList(((org.bukkit.inventory.InventoryHolder) p_143375_.getBukkitEntity()).getInventory().getViewers())) {
+                    h.closeInventory(org.bukkit.event.inventory.InventoryCloseEvent.Reason.UNLOADED); // Paper - Inventory close reason
+                }
+            }
+            // Spigot End
             ServerLevel.this.getChunkSource().removeEntity(p_143375_);
             if (p_143375_ instanceof ServerPlayer serverplayer) {
                 ServerLevel.this.players.remove(serverplayer);
@@ -1733,24 +_,121 @@
                 if (ServerLevel.this.isUpdatingNavigations) {
                     String s = "onTrackingStart called during navigation iteration";
                     Util.logAndPauseIfInIde(
-                        "onTrackingStart called during navigation iteration", new IllegalStateException("onTrackingStart called during navigation iteration")
+                              "onTrackingStart called during navigation iteration", new IllegalStateException("onTrackingStart called during navigation iteration")
                     );
                 }
 
                 ServerLevel.this.navigatingMobs.remove(mob);
             }
 
-            if (p_143375_ instanceof EnderDragon enderdragon) {
-                for (EnderDragonPart enderdragonpart : enderdragon.getSubEntities()) {
+            if (p_143375_.isMultipartEntity()) {
+                for(net.neoforged.neoforge.entity.PartEntity<?> enderdragonpart : p_143375_.getParts()) {
                     ServerLevel.this.dragonParts.remove(enderdragonpart.getId());
                 }
             }
 
             p_143375_.updateDynamicGameEventListener(DynamicGameEventListener::remove);
+
+            // CraftBukkit start
+            p_143375_.valid = false;
+            if (!(p_143375_ instanceof ServerPlayer)) {
+                for (ServerPlayer player : players) {
+                    player.getBukkitEntity().onEntityRemove(p_143375_);
+                }
+            }
+            // CraftBukkit end
+
+            p_143375_.onRemovedFromLevel();
+            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.EntityLeaveLevelEvent(p_143375_, ServerLevel.this));
+            new com.destroystokyo.paper.event.entity.EntityRemoveFromWorldEvent(p_143375_.getBukkitEntity(), ServerLevel.this.getWorld()).callEvent(); // Paper - fire while valid
         }
 
         public void onSectionChange(Entity p_215086_) {
             p_215086_.updateDynamicGameEventListener(DynamicGameEventListener::move);
         }
     }
+
+    @Override
+    public java.util.Collection<net.neoforged.neoforge.entity.PartEntity<?>> getPartEntities() {
+        return this.dragonParts.values();
+    }
+
+    @Override
+    public final void syncData(net.neoforged.neoforge.attachment.AttachmentType<?> type) {
+        net.neoforged.neoforge.attachment.AttachmentSync.syncLevelUpdate(this, type);
+    }
+
+    private final net.neoforged.neoforge.capabilities.CapabilityListenerHolder capListenerHolder = new net.neoforged.neoforge.capabilities.CapabilityListenerHolder();
+
+    @Override
+    public void invalidateCapabilities(BlockPos pos) {
+        capListenerHolder.invalidatePos(pos);
+    }
+
+    @Override
+    public void invalidateCapabilities(ChunkPos pos) {
+        capListenerHolder.invalidateChunk(pos);
+    }
+
+    /**
+     * Register a listener for capability invalidation.
+     * @see net.neoforged.neoforge.capabilities.ICapabilityInvalidationListener
+     */
+    public void registerCapabilityListener(BlockPos pos, net.neoforged.neoforge.capabilities.ICapabilityInvalidationListener listener) {
+        capListenerHolder.addListener(pos, listener);
+    }
+
+    /**
+     * Internal method, used to clean capability listeners that are not referenced.
+     * Do not call.
+     */
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public void cleanCapabilityListenerReferences() {
+        capListenerHolder.clean();
+    }
+
+    // Neo: Variable day time code
+
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public void setDayTimeFraction(float dayTimeFraction) {
+        serverLevelData.setDayTimeFraction(dayTimeFraction);
+    }
+
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public float getDayTimeFraction() {
+        return serverLevelData.getDayTimeFraction();
+    }
+
+    /**
+     * Returns the current ratio between game ticks and clock ticks. If this value is negative, no
+     * speed has been set and those two are coupled 1:1 (i.e. vanilla mode).
+     */
+    public float getDayTimePerTick() {
+        return serverLevelData.getDayTimePerTick();
+    }
+
+    /**
+     * This allows mods to set the rate time flows in a level. By default, each game tick the clock time
+     * also advances by one tick, with {@link Level#TICKS_PER_DAY} clock ticks (or 20 real-life minutes)
+     * forming a Minecraft day.
+     * <p>
+     * This can be sped up for shorter days by giving a higher number, or slowed down for longer days
+     * with a smaller number. A negative value will reset it back to vanilla logic.
+     * <p>
+     * This value can also be changed with the command <code>/neoforge day</code>, where you can set
+     * either the speed or a day length in minutes.
+     * <p>
+     * This has no effect when time progression is stopped.
+     * <p>
+     * While this still technically works when vanilla clients are connected, those will desync and
+     * experience a time jump once per second.
+     */
+    @Override
+    public void setDayTimePerTick(float dayTimePerTick) {
+        if (dayTimePerTick != getDayTimePerTick() && dayTimePerTick != 0f) {
+            serverLevelData.setDayTimePerTick(dayTimePerTick);
+            server.forceTimeSynchronization();
+        }
+    }
+
 }
