--- a/net/minecraft/server/commands/KickCommand.java
+++ b/net/minecraft/server/commands/KickCommand.java
@@ -53,7 +_,7 @@
 
             for (ServerPlayer serverplayer : p_137803_) {
                 if (!p_137802_.getServer().isSingleplayerOwner(serverplayer.getGameProfile())) {
-                    serverplayer.connection.disconnect(p_137804_);
+                    serverplayer.connection.disconnect(p_137804_, org.bukkit.event.player.PlayerKickEvent.Cause.KICK_COMMAND); // Paper - kick event cause
                     p_137802_.sendSuccess(() -> Component.translatable("commands.kick.success", serverplayer.getDisplayName(), p_137804_), true);
                     i++;
                 }
