--- a/net/minecraft/server/commands/SummonCommand.java
+++ b/net/minecraft/server/commands/SummonCommand.java
@@ -85,6 +_,7 @@
             ServerLevel serverlevel = p_270582_.getLevel();
             Entity entity = EntityType.loadEntityRecursive(compoundtag, serverlevel, p_138828_ -> {
                 p_138828_.moveTo(p_270366_.x, p_270366_.y, p_270366_.z, p_138828_.getYRot(), p_138828_.getXRot());
+                p_138828_.spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.COMMAND; // Paper - Entity#getEntitySpawnReason
                 return p_138828_;
             });
             if (entity == null) {
@@ -95,7 +_,7 @@
                         .finalizeSpawn(p_270582_.getLevel(), p_270582_.getLevel().getCurrentDifficultyAt(entity.blockPosition()), MobSpawnType.COMMAND, null);
                 }
 
-                if (!serverlevel.tryAddFreshEntityWithPassengers(entity)) {
+                if (!serverlevel.tryAddFreshEntityWithPassengers(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.COMMAND)) { // CraftBukkit - pass a spawn reason of "COMMAND"
                     throw ERROR_DUPLICATE_UUID.create();
                 } else {
                     return entity;
